{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ba38843d-12ba-4585-b30f-e9ac08b987aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/cstansbu/miniconda3/envs/bioinf/lib/python3.10/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(57180, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000104478</td>\n",
       "      <td>2</td>\n",
       "      <td>Gm38212</td>\n",
       "      <td>havana</td>\n",
       "      <td>TEC</td>\n",
       "      <td>1</td>\n",
       "      <td>108344806</td>\n",
       "      <td>108347562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000104385</td>\n",
       "      <td>2</td>\n",
       "      <td>Gm7449</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>6980783</td>\n",
       "      <td>6981446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000102135</td>\n",
       "      <td>2</td>\n",
       "      <td>Gm37108</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>6986782</td>\n",
       "      <td>6993812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSMUSG00000103282</td>\n",
       "      <td>2</td>\n",
       "      <td>Gm37275</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>6999982</td>\n",
       "      <td>7000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSMUSG00000101097</td>\n",
       "      <td>2</td>\n",
       "      <td>Gm6679</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>108697864</td>\n",
       "      <td>108699733</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id gene_version gene_name gene_source  \\\n",
       "0   ENSMUSG00000104478            2   Gm38212      havana   \n",
       "3   ENSMUSG00000104385            2    Gm7449      havana   \n",
       "6   ENSMUSG00000102135            2   Gm37108      havana   \n",
       "10  ENSMUSG00000103282            2   Gm37275      havana   \n",
       "13  ENSMUSG00000101097            2    Gm6679      havana   \n",
       "\n",
       "            gene_biotype Chromosome      Start        End  \n",
       "0                    TEC          1  108344806  108347562  \n",
       "3   processed_pseudogene          1    6980783    6981446  \n",
       "6   processed_pseudogene          1    6986782    6993812  \n",
       "10  processed_pseudogene          1    6999982    7000012  \n",
       "13  processed_pseudogene          1  108697864  108699733  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "keep_columns = [\n",
    "    'gene_id', \n",
    "    'gene_version', \n",
    "    'gene_name', \n",
    "    'gene_source',\n",
    "    'gene_biotype', \n",
    "    'Chromosome',\n",
    "    'Start',\n",
    "    'End',\n",
    "]\n",
    "\n",
    "fpath = \"/scratch/indikar_root/indikar1/shared_data/scpore_c/gtf/GRCm39.gtf.gz\"\n",
    "df = pr.read_gtf(fpath).as_df()\n",
    "\n",
    "df = df[df['Feature'] == 'gene']\n",
    "df = df[keep_columns]\n",
    "df = df.drop_duplicates()\n",
    "print(df.shape)\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d6e87716-ebde-41df-9d82-255ae39b1381",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(56655, 11)\n",
      "1076\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>length</th>\n",
       "      <th>midpoint</th>\n",
       "      <th>is_tf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000104478</td>\n",
       "      <td>2</td>\n",
       "      <td>Gm38212</td>\n",
       "      <td>havana</td>\n",
       "      <td>TEC</td>\n",
       "      <td>1</td>\n",
       "      <td>108344806</td>\n",
       "      <td>108347562</td>\n",
       "      <td>2756</td>\n",
       "      <td>108346184</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000104385</td>\n",
       "      <td>2</td>\n",
       "      <td>Gm7449</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>6980783</td>\n",
       "      <td>6981446</td>\n",
       "      <td>663</td>\n",
       "      <td>6981114</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000102135</td>\n",
       "      <td>2</td>\n",
       "      <td>Gm37108</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>6986782</td>\n",
       "      <td>6993812</td>\n",
       "      <td>7030</td>\n",
       "      <td>6990297</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000103282</td>\n",
       "      <td>2</td>\n",
       "      <td>Gm37275</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>6999982</td>\n",
       "      <td>7000012</td>\n",
       "      <td>30</td>\n",
       "      <td>6999997</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000101097</td>\n",
       "      <td>2</td>\n",
       "      <td>Gm6679</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>108697864</td>\n",
       "      <td>108699733</td>\n",
       "      <td>1869</td>\n",
       "      <td>108698798</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id gene_version gene_name gene_source  \\\n",
       "0  ENSMUSG00000104478            2   Gm38212      havana   \n",
       "1  ENSMUSG00000104385            2    Gm7449      havana   \n",
       "2  ENSMUSG00000102135            2   Gm37108      havana   \n",
       "3  ENSMUSG00000103282            2   Gm37275      havana   \n",
       "4  ENSMUSG00000101097            2    Gm6679      havana   \n",
       "\n",
       "           gene_biotype Chromosome      Start        End  length   midpoint  \\\n",
       "0                   TEC          1  108344806  108347562    2756  108346184   \n",
       "1  processed_pseudogene          1    6980783    6981446     663    6981114   \n",
       "2  processed_pseudogene          1    6986782    6993812    7030    6990297   \n",
       "3  processed_pseudogene          1    6999982    7000012      30    6999997   \n",
       "4  processed_pseudogene          1  108697864  108699733    1869  108698798   \n",
       "\n",
       "   is_tf  \n",
       "0  False  \n",
       "1  False  \n",
       "2  False  \n",
       "3  False  \n",
       "4  False  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = \"/scratch/indikar_root/indikar1/shared_data/higher_order/reference/gene_table.parquet\"\n",
    "df = pd.read_parquet(fpath)\n",
    "print(df.shape)\n",
    "print(df['is_tf'].sum())\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0ca58edf-a553-497d-b6d3-1c8f54b238c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chromosome\n",
       "2             106\n",
       "7              80\n",
       "11             78\n",
       "1              70\n",
       "4              67\n",
       "5              62\n",
       "10             61\n",
       "6              60\n",
       "8              56\n",
       "9              48\n",
       "15             47\n",
       "12             46\n",
       "3              45\n",
       "13             41\n",
       "17             40\n",
       "19             39\n",
       "14             36\n",
       "X              35\n",
       "16             29\n",
       "18             29\n",
       "Y               1\n",
       "GL456210.1      0\n",
       "GL456212.1      0\n",
       "GL456211.1      0\n",
       "GL456219.1      0\n",
       "GL456221.1      0\n",
       "GL456372.1      0\n",
       "GL456381.1      0\n",
       "GL456239.1      0\n",
       "GL456354.1      0\n",
       "JH584296.1      0\n",
       "GL456385.1      0\n",
       "JH584297.1      0\n",
       "JH584298.1      0\n",
       "JH584303.1      0\n",
       "JH584299.1      0\n",
       "MT              0\n",
       "JH584304.1      0\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['is_tf']]['Chromosome'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4d8c6a9d-6d7e-410d-b621-d7bdbfff0571",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ABL1</th>\n",
       "      <th>ACAA1</th>\n",
       "      <th>ADNP</th>\n",
       "      <th>ADNP2</th>\n",
       "      <th>AEBP2</th>\n",
       "      <th>AFF4</th>\n",
       "      <th>AHCTF1</th>\n",
       "      <th>AHDC1</th>\n",
       "      <th>AHR</th>\n",
       "      <th>AHRR</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSCAN32</th>\n",
       "      <th>ZSCAN4</th>\n",
       "      <th>ZSCAN5A</th>\n",
       "      <th>ZSCAN5B</th>\n",
       "      <th>ZSCAN5C</th>\n",
       "      <th>ZSCAN9</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.300</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.25</td>\n",
       "      <td>4.37</td>\n",
       "      <td>0.018250</td>\n",
       "      <td>1.2576</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.538</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <td>6.56</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.685</td>\n",
       "      <td>0.841</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.55</td>\n",
       "      <td>1.54</td>\n",
       "      <td>0.449875</td>\n",
       "      <td>1.0866</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.4870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>1.18</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.721</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.13</td>\n",
       "      <td>1.44</td>\n",
       "      <td>0.282500</td>\n",
       "      <td>1.4038</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.8900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.414</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.24</td>\n",
       "      <td>1.43</td>\n",
       "      <td>1.109500</td>\n",
       "      <td>2.3160</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009O20Rik</th>\n",
       "      <td>2.10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.147</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.12</td>\n",
       "      <td>1.75</td>\n",
       "      <td>0.471513</td>\n",
       "      <td>0.1480</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2870</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1605 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               ABL1  ACAA1   ADNP  ADNP2  AEBP2  AFF4  AHCTF1     AHDC1  \\\n",
       "gene_name                                                                 \n",
       "0610005C13Rik  0.00    0.0  4.300  0.000    0.0  5.25    4.37  0.018250   \n",
       "0610007P14Rik  6.56    0.0  0.685  0.841    0.0  2.55    1.54  0.449875   \n",
       "0610009B22Rik  1.18    0.0  0.721  0.000    0.0  1.13    1.44  0.282500   \n",
       "0610009L18Rik  0.00    0.0  0.000  0.414    0.0  1.24    1.43  1.109500   \n",
       "0610009O20Rik  2.10    0.0  0.147  0.000    0.0  5.12    1.75  0.471513   \n",
       "\n",
       "                  AHR  AHRR  ...  ZSCAN32  ZSCAN4  ZSCAN5A  ZSCAN5B  ZSCAN5C  \\\n",
       "gene_name                    ...                                               \n",
       "0610005C13Rik  1.2576   0.0  ...      0.0     0.0      0.0      0.0    0.538   \n",
       "0610007P14Rik  1.0866   0.0  ...      0.0     0.0      0.0      0.0    0.000   \n",
       "0610009B22Rik  1.4038   0.0  ...      0.0     0.0      0.0      0.0    0.000   \n",
       "0610009L18Rik  2.3160   0.0  ...      0.0     0.0      0.0      0.0    0.000   \n",
       "0610009O20Rik  0.1480   0.0  ...      0.0     0.0      0.0      0.0    0.000   \n",
       "\n",
       "               ZSCAN9  ZXDA  ZXDB  ZXDC    ZZZ3  \n",
       "gene_name                                        \n",
       "0610005C13Rik     0.0   0.0   0.0   0.0  0.0326  \n",
       "0610007P14Rik     0.0   0.0   0.0   0.0  0.4870  \n",
       "0610009B22Rik     0.0   0.0   0.0   0.0  1.8900  \n",
       "0610009L18Rik     0.0   0.0   0.0   0.0  0.0000  \n",
       "0610009O20Rik     0.0   0.0   0.0   0.0  0.2870  \n",
       "\n",
       "[5 rows x 1605 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scenic_path = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/processed_data/MOUSE_500bp_up_100bp_down_B.csv\"\n",
    "\n",
    "df = pd.read_csv(scenic_path)\n",
    "df = df.rename(columns={'Unnamed: 0' : 'gene_name'})\n",
    "df = df.set_index('gene_name')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f44c02bf-287d-4a6e-be7e-87cf0ee04f8a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinf",
   "language": "python",
   "name": "bioinf"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
