{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5b5b9714-eab5-4e83-94a0-dc7e6a674e6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import time\n",
    "import gget\n",
    "import scipy\n",
    "from scipy.sparse import csr_matrix\n",
    "import anndata as an\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import matplotlib.patches as mpatches\n",
    "import networkx as nx\n",
    "import random\n",
    "from importlib import reload\n",
    "import warnings\n",
    "import ot\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from matplotlib.colors import ListedColormap\n",
    "\n",
    "import surprise as sup\n",
    "\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "\n",
    "\"\"\"WARNING: no warnings\"\"\"\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# local imports\n",
    "import anndata_utils as anntools\n",
    "\n",
    "source_path = os.path.abspath(\"../source/\")\n",
    "sys.path.append(source_path)\n",
    "import centrality as central\n",
    "import matrix\n",
    "import utils as ut\n",
    "import plotting as plt2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8fd2c2ba-feb8-4585-bee2-57e814f06cfa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time taken to read the file: 6.96 seconds\n",
      "Memory usage: current 1.53 GB, difference +1.53 GB\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 2579 × 2756467\n",
       "    obs: 'bin_index', 'bin_start', 'bin_end', 'bin', 'chrom', 'chrom_bin', 'degree', 'genes', 'n_genes', 'ATACSeq_1', 'ATACSeq_2', 'ATACSeq_3', 'CTCF', 'H3K27ac', 'H3K27me3', 'RNA_1', 'RNA_2', 'RNA_3', 'RNA_4', 'RNA_5', 'RNA_6'\n",
       "    var: 'read_index', 'basename', 'mean_mapq', 'median_mapq', 'n_chromosomes', 'order', 'n_bins', 'read_length_bp', 'genes', 'n_genes'\n",
       "    uns: 'base_resolution', 'chrom_sizes'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load ann data\n",
    "resolution = 1000000\n",
    "population_path = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/lightweight/population_mESC_{resolution}_anndata.h5ad\"\n",
    "\n",
    "start_time = time.time()  # Record the start time\n",
    "adata = sc.read_h5ad(population_path)\n",
    "end_time = time.time()  # Record the end time\n",
    "print(f\"Time taken to read the file: {end_time - start_time:.2f} seconds\")\n",
    "sc.logging.print_memory_usage()\n",
    "\n",
    "# sort the object by global bin id\n",
    "sorted_indices = adata.obs['bin'].argsort()\n",
    "adata = adata[sorted_indices, :] \n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "16e29329-9836-4e37-ad92-fe141acc80c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_index</th>\n",
       "      <th>bin_start</th>\n",
       "      <th>bin_end</th>\n",
       "      <th>bin</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chrom_bin</th>\n",
       "      <th>degree</th>\n",
       "      <th>genes</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>ATACSeq_1</th>\n",
       "      <th>...</th>\n",
       "      <th>ATACSeq_3</th>\n",
       "      <th>CTCF</th>\n",
       "      <th>H3K27ac</th>\n",
       "      <th>H3K27me3</th>\n",
       "      <th>RNA_1</th>\n",
       "      <th>RNA_2</th>\n",
       "      <th>RNA_3</th>\n",
       "      <th>RNA_4</th>\n",
       "      <th>RNA_5</th>\n",
       "      <th>RNA_6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bin_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:3</th>\n",
       "      <td>2235</td>\n",
       "      <td>3000000</td>\n",
       "      <td>4000000</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2953</td>\n",
       "      <td>Gm37363;Gm37381;Gm7341;Gm38148;Gm37180;Gm19938...</td>\n",
       "      <td>15</td>\n",
       "      <td>0.507852</td>\n",
       "      <td>...</td>\n",
       "      <td>0.375103</td>\n",
       "      <td>0.319337</td>\n",
       "      <td>0.211256</td>\n",
       "      <td>0.577305</td>\n",
       "      <td>0.263153</td>\n",
       "      <td>0.203017</td>\n",
       "      <td>0.128587</td>\n",
       "      <td>0.271721</td>\n",
       "      <td>0.124134</td>\n",
       "      <td>0.093115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:4</th>\n",
       "      <td>964</td>\n",
       "      <td>4000000</td>\n",
       "      <td>5000000</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>3331</td>\n",
       "      <td>Gm7369;A930006A01Rik;Gm37381;Gm38076;Mrpl15;Gm...</td>\n",
       "      <td>22</td>\n",
       "      <td>0.680719</td>\n",
       "      <td>...</td>\n",
       "      <td>0.586783</td>\n",
       "      <td>0.561063</td>\n",
       "      <td>0.698206</td>\n",
       "      <td>0.735207</td>\n",
       "      <td>0.282745</td>\n",
       "      <td>0.706667</td>\n",
       "      <td>0.265677</td>\n",
       "      <td>0.247106</td>\n",
       "      <td>0.684904</td>\n",
       "      <td>0.225147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:5</th>\n",
       "      <td>1823</td>\n",
       "      <td>5000000</td>\n",
       "      <td>6000000</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>3341</td>\n",
       "      <td>Atp6v1h;Gm17101;Gm16041;Rgs20;Gm36965;Gm38264;...</td>\n",
       "      <td>12</td>\n",
       "      <td>0.567134</td>\n",
       "      <td>...</td>\n",
       "      <td>0.467917</td>\n",
       "      <td>0.426098</td>\n",
       "      <td>0.410865</td>\n",
       "      <td>0.458616</td>\n",
       "      <td>0.197587</td>\n",
       "      <td>0.273313</td>\n",
       "      <td>0.233012</td>\n",
       "      <td>0.157297</td>\n",
       "      <td>0.282102</td>\n",
       "      <td>0.108284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:6</th>\n",
       "      <td>1288</td>\n",
       "      <td>6000000</td>\n",
       "      <td>7000000</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>3422</td>\n",
       "      <td>St18;Gm7449;Gm19214;Gm37108;Gm2147;Gm19026;473...</td>\n",
       "      <td>10</td>\n",
       "      <td>0.621495</td>\n",
       "      <td>...</td>\n",
       "      <td>0.493767</td>\n",
       "      <td>0.474813</td>\n",
       "      <td>0.412592</td>\n",
       "      <td>0.640384</td>\n",
       "      <td>0.209920</td>\n",
       "      <td>0.285676</td>\n",
       "      <td>0.147348</td>\n",
       "      <td>0.170128</td>\n",
       "      <td>0.248226</td>\n",
       "      <td>0.116792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7</th>\n",
       "      <td>144</td>\n",
       "      <td>7000000</td>\n",
       "      <td>8000000</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>3289</td>\n",
       "      <td>Gm23274;Gm5694;Rps2-ps2;Gm37225;Gm37489;Gm2690...</td>\n",
       "      <td>18</td>\n",
       "      <td>0.497991</td>\n",
       "      <td>...</td>\n",
       "      <td>0.427530</td>\n",
       "      <td>0.387846</td>\n",
       "      <td>0.458324</td>\n",
       "      <td>0.391387</td>\n",
       "      <td>0.229293</td>\n",
       "      <td>0.227452</td>\n",
       "      <td>0.157668</td>\n",
       "      <td>0.166332</td>\n",
       "      <td>0.212388</td>\n",
       "      <td>0.116124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          bin_index  bin_start  bin_end  bin chrom  chrom_bin  degree  \\\n",
       "bin_name                                                                \n",
       "chr1:3         2235    3000000  4000000    3     1          3    2953   \n",
       "chr1:4          964    4000000  5000000    4     1          4    3331   \n",
       "chr1:5         1823    5000000  6000000    5     1          5    3341   \n",
       "chr1:6         1288    6000000  7000000    6     1          6    3422   \n",
       "chr1:7          144    7000000  8000000    7     1          7    3289   \n",
       "\n",
       "                                                      genes  n_genes  \\\n",
       "bin_name                                                               \n",
       "chr1:3    Gm37363;Gm37381;Gm7341;Gm38148;Gm37180;Gm19938...       15   \n",
       "chr1:4    Gm7369;A930006A01Rik;Gm37381;Gm38076;Mrpl15;Gm...       22   \n",
       "chr1:5    Atp6v1h;Gm17101;Gm16041;Rgs20;Gm36965;Gm38264;...       12   \n",
       "chr1:6    St18;Gm7449;Gm19214;Gm37108;Gm2147;Gm19026;473...       10   \n",
       "chr1:7    Gm23274;Gm5694;Rps2-ps2;Gm37225;Gm37489;Gm2690...       18   \n",
       "\n",
       "          ATACSeq_1  ...  ATACSeq_3      CTCF   H3K27ac  H3K27me3     RNA_1  \\\n",
       "bin_name             ...                                                      \n",
       "chr1:3     0.507852  ...   0.375103  0.319337  0.211256  0.577305  0.263153   \n",
       "chr1:4     0.680719  ...   0.586783  0.561063  0.698206  0.735207  0.282745   \n",
       "chr1:5     0.567134  ...   0.467917  0.426098  0.410865  0.458616  0.197587   \n",
       "chr1:6     0.621495  ...   0.493767  0.474813  0.412592  0.640384  0.209920   \n",
       "chr1:7     0.497991  ...   0.427530  0.387846  0.458324  0.391387  0.229293   \n",
       "\n",
       "             RNA_2     RNA_3     RNA_4     RNA_5     RNA_6  \n",
       "bin_name                                                    \n",
       "chr1:3    0.203017  0.128587  0.271721  0.124134  0.093115  \n",
       "chr1:4    0.706667  0.265677  0.247106  0.684904  0.225147  \n",
       "chr1:5    0.273313  0.233012  0.157297  0.282102  0.108284  \n",
       "chr1:6    0.285676  0.147348  0.170128  0.248226  0.116792  \n",
       "chr1:7    0.227452  0.157668  0.166332  0.212388  0.116124  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5305b34b-5d96-4405-87dd-cd5720d8c9df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "degree_outlier\n",
      "False    2431\n",
      "True      148\n",
      "Name: count, dtype: int64\n",
      "Removing top 148 outlier loci: \n",
      "['chr1:24', 'chr1:85', 'chr1:195', 'chr2:98', 'chr2:177', 'chr4:112', 'chr4:145', 'chr4:146', 'chr4:147', 'chr5:94', 'chr5:95', 'chr6:103', 'chr7:20', 'chr7:21', 'chr7:22', 'chr7:32', 'chr8:19', 'chr8:20', 'chr8:21', 'chr8:70', 'chr8:130', 'chr9:3', 'chr9:35', 'chr9:124', 'chr10:130', 'chr11:3', 'chr12:3', 'chr13:65', 'chr13:66', 'chr14:3', 'chr14:15', 'chr14:16', 'chr14:17', 'chr14:18', 'chr14:19', 'chr14:125', 'chr16:3', 'chr17:95', 'chr19:61', 'chrX:6', 'chrX:8', 'chrX:9', 'chrX:10', 'chrX:14', 'chrX:15', 'chrX:16', 'chrX:17', 'chrX:18', 'chrX:19', 'chrX:22', 'chrX:23', 'chrX:24', 'chrX:25', 'chrX:26', 'chrX:27', 'chrX:29', 'chrX:34', 'chrX:38', 'chrX:39', 'chrX:40', 'chrX:42', 'chrX:43', 'chrX:44', 'chrX:45', 'chrX:46', 'chrX:48', 'chrX:49', 'chrX:53', 'chrX:54', 'chrX:57', 'chrX:58', 'chrX:60', 'chrX:61', 'chrX:62', 'chrX:63', 'chrX:64', 'chrX:65', 'chrX:66', 'chrX:67', 'chrX:68', 'chrX:69', 'chrX:71', 'chrX:77', 'chrX:78', 'chrX:79', 'chrX:80', 'chrX:81', 'chrX:82', 'chrX:83', 'chrX:84', 'chrX:85', 'chrX:86', 'chrX:87', 'chrX:88', 'chrX:89', 'chrX:90', 'chrX:91', 'chrX:93', 'chrX:94', 'chrX:95', 'chrX:96', 'chrX:97', 'chrX:98', 'chrX:104', 'chrX:106', 'chrX:107', 'chrX:108', 'chrX:109', 'chrX:110', 'chrX:111', 'chrX:112', 'chrX:113', 'chrX:114', 'chrX:115', 'chrX:116', 'chrX:117', 'chrX:118', 'chrX:119', 'chrX:120', 'chrX:121', 'chrX:122', 'chrX:123', 'chrX:124', 'chrX:125', 'chrX:126', 'chrX:127', 'chrX:128', 'chrX:130', 'chrX:131', 'chrX:132', 'chrX:134', 'chrX:135', 'chrX:136', 'chrX:137', 'chrX:139', 'chrX:140', 'chrX:141', 'chrX:144', 'chrX:145', 'chrX:154', 'chrX:155', 'chrX:156', 'chrX:157', 'chrX:160', 'chrX:164', 'chrX:166', 'chrX:167', 'chrX:169']\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "def find_outliers(df_column):\n",
    "  \"\"\"\n",
    "  Identifies outliers in a pandas DataFrame column using the IQR method.\n",
    "\n",
    "  Args:\n",
    "    df_column: A pandas Series representing the column to analyze.\n",
    "\n",
    "  Returns:\n",
    "    A boolean mask with True for outliers and False otherwise.\n",
    "  \"\"\"\n",
    "  Q1 = df_column.quantile(0.15)\n",
    "  Q3 = df_column.quantile(0.85)\n",
    "  IQR = Q3 - Q1\n",
    "  lower_bound = Q1 - 1.5 * IQR\n",
    "  upper_bound = Q3 + 1.5 * IQR\n",
    "  return (df_column < lower_bound) | (df_column > upper_bound)\n",
    "\n",
    "adata.obs['degree_outlier'] = find_outliers(adata.obs['degree'])\n",
    "\n",
    "print(adata.obs['degree_outlier'].value_counts())\n",
    "\n",
    "# remove outliers\n",
    "remove_bins = adata.obs[adata.obs['degree_outlier']].index.to_list()\n",
    "print(f\"Removing top {len(remove_bins)} outlier loci: \")\n",
    "print(remove_bins)\n",
    "\n",
    "adata = adata[~adata.obs_names.isin(remove_bins), :].copy()\n",
    "\n",
    "print('done!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4f17f66a-1528-429b-a541-9c5c8d963ffb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the core scores "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b98d9a16-1df6-46a8-8b1f-dbbe5e52071a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scores.shape=(2431, 36)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_name</th>\n",
       "      <th>bin_index</th>\n",
       "      <th>bin_start</th>\n",
       "      <th>bin_end</th>\n",
       "      <th>bin</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chrom_bin</th>\n",
       "      <th>degree</th>\n",
       "      <th>genes</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>...</th>\n",
       "      <th>ce_betweenness_centrality</th>\n",
       "      <th>ce_pagerank</th>\n",
       "      <th>hge_singular_vector_1</th>\n",
       "      <th>hge_logexp_unweighted</th>\n",
       "      <th>hge_logexp_degree_weighted</th>\n",
       "      <th>hge_logexp_RNA_weighted</th>\n",
       "      <th>hge_logexp_ATAC_weighted</th>\n",
       "      <th>global_singular_vector_1</th>\n",
       "      <th>global_hge_logexp_unweighted</th>\n",
       "      <th>global_hge_logexp_RNA_weighted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:3</td>\n",
       "      <td>2235</td>\n",
       "      <td>3000000</td>\n",
       "      <td>4000000</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2953</td>\n",
       "      <td>Gm37363;Gm37381;Gm7341;Gm38148;Gm37180;Gm19938...</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>0.671480</td>\n",
       "      <td>0.117995</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.148099</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.174370</td>\n",
       "      <td>0.222380</td>\n",
       "      <td>0.330855</td>\n",
       "      <td>0.501567</td>\n",
       "      <td>0.167764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:4</td>\n",
       "      <td>964</td>\n",
       "      <td>4000000</td>\n",
       "      <td>5000000</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>3331</td>\n",
       "      <td>Gm7369;A930006A01Rik;Gm37381;Gm38076;Mrpl15;Gm...</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>0.267148</td>\n",
       "      <td>0.372869</td>\n",
       "      <td>0.103845</td>\n",
       "      <td>0.239187</td>\n",
       "      <td>0.281148</td>\n",
       "      <td>0.468270</td>\n",
       "      <td>0.560302</td>\n",
       "      <td>0.412235</td>\n",
       "      <td>0.609261</td>\n",
       "      <td>0.369136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:5</td>\n",
       "      <td>1823</td>\n",
       "      <td>5000000</td>\n",
       "      <td>6000000</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>3341</td>\n",
       "      <td>Atp6v1h;Gm17101;Gm16041;Rgs20;Gm36965;Gm38264;...</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>0.109507</td>\n",
       "      <td>0.224249</td>\n",
       "      <td>0.057865</td>\n",
       "      <td>0.647027</td>\n",
       "      <td>0.201096</td>\n",
       "      <td>0.243125</td>\n",
       "      <td>0.405148</td>\n",
       "      <td>0.425601</td>\n",
       "      <td>0.528269</td>\n",
       "      <td>0.216383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:6</td>\n",
       "      <td>1288</td>\n",
       "      <td>6000000</td>\n",
       "      <td>7000000</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>3422</td>\n",
       "      <td>St18;Gm7449;Gm19214;Gm37108;Gm2147;Gm19026;473...</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009627</td>\n",
       "      <td>0.534940</td>\n",
       "      <td>0.172229</td>\n",
       "      <td>0.760979</td>\n",
       "      <td>0.401754</td>\n",
       "      <td>0.250762</td>\n",
       "      <td>0.521777</td>\n",
       "      <td>0.448264</td>\n",
       "      <td>0.492633</td>\n",
       "      <td>0.258762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:7</td>\n",
       "      <td>144</td>\n",
       "      <td>7000000</td>\n",
       "      <td>8000000</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>3289</td>\n",
       "      <td>Gm23274;Gm5694;Rps2-ps2;Gm37225;Gm37489;Gm2690...</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>0.252708</td>\n",
       "      <td>0.099600</td>\n",
       "      <td>0.053319</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.160235</td>\n",
       "      <td>0.190846</td>\n",
       "      <td>0.204818</td>\n",
       "      <td>0.414267</td>\n",
       "      <td>0.510526</td>\n",
       "      <td>0.204855</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  bin_name  bin_index  bin_start  bin_end  bin chrom  chrom_bin  degree  \\\n",
       "0   chr1:3       2235    3000000  4000000    3     1          3    2953   \n",
       "1   chr1:4        964    4000000  5000000    4     1          4    3331   \n",
       "2   chr1:5       1823    5000000  6000000    5     1          5    3341   \n",
       "3   chr1:6       1288    6000000  7000000    6     1          6    3422   \n",
       "4   chr1:7        144    7000000  8000000    7     1          7    3289   \n",
       "\n",
       "                                               genes  n_genes  ...  \\\n",
       "0  Gm37363;Gm37381;Gm7341;Gm38148;Gm37180;Gm19938...       15  ...   \n",
       "1  Gm7369;A930006A01Rik;Gm37381;Gm38076;Mrpl15;Gm...       22  ...   \n",
       "2  Atp6v1h;Gm17101;Gm16041;Rgs20;Gm36965;Gm38264;...       12  ...   \n",
       "3  St18;Gm7449;Gm19214;Gm37108;Gm2147;Gm19026;473...       10  ...   \n",
       "4  Gm23274;Gm5694;Rps2-ps2;Gm37225;Gm37489;Gm2690...       18  ...   \n",
       "\n",
       "   ce_betweenness_centrality  ce_pagerank  hge_singular_vector_1  \\\n",
       "0                   0.671480     0.117995               0.000000   \n",
       "1                   0.267148     0.372869               0.103845   \n",
       "2                   0.109507     0.224249               0.057865   \n",
       "3                   0.009627     0.534940               0.172229   \n",
       "4                   0.252708     0.099600               0.053319   \n",
       "\n",
       "   hge_logexp_unweighted  hge_logexp_degree_weighted  hge_logexp_RNA_weighted  \\\n",
       "0               0.148099                    0.000000                 0.174370   \n",
       "1               0.239187                    0.281148                 0.468270   \n",
       "2               0.647027                    0.201096                 0.243125   \n",
       "3               0.760979                    0.401754                 0.250762   \n",
       "4               0.000000                    0.160235                 0.190846   \n",
       "\n",
       "   hge_logexp_ATAC_weighted  global_singular_vector_1  \\\n",
       "0                  0.222380                  0.330855   \n",
       "1                  0.560302                  0.412235   \n",
       "2                  0.405148                  0.425601   \n",
       "3                  0.521777                  0.448264   \n",
       "4                  0.204818                  0.414267   \n",
       "\n",
       "   global_hge_logexp_unweighted  global_hge_logexp_RNA_weighted  \n",
       "0                      0.501567                        0.167764  \n",
       "1                      0.609261                        0.369136  \n",
       "2                      0.528269                        0.216383  \n",
       "3                      0.492633                        0.258762  \n",
       "4                      0.510526                        0.204855  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = \"/scratch/indikar_root/indikar1/shared_data/higher_order/global_core_score/population_mESC_1000000_scores.csv\"\n",
    "scores = pd.read_csv(fpath)\n",
    "print(f\"{scores.shape=}\")\n",
    "scores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cf11e769-cb5a-44b8-913b-92b2602ab4b7",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'break' outside loop (668683560.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[7], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    break\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m 'break' outside loop\n"
     ]
    }
   ],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "772763f8-dd41-4f76-892b-292cefb00297",
   "metadata": {},
   "source": [
    "# load the core scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39553852-6941-41cb-b5ac-07900f9660cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "dpath = \"/scratch/indikar_root/indikar1/shared_data/higher_order/core_scores/\"\n",
    "\n",
    "file_list = glob.glob(f\"{dpath}population_mESC_{resolution}*\")\n",
    "print(f\"{len(file_list)=}\")\n",
    "\n",
    "\n",
    "df = []\n",
    "\n",
    "for fpath in file_list:\n",
    "    basename = os.path.basename(fpath)\n",
    "    scores = pd.read_csv(fpath)\n",
    "    print(basename, scores.shape)\n",
    "    df.append(scores)\n",
    "\n",
    "\n",
    "df = pd.concat(df)\n",
    "\n",
    "# drop outliers \n",
    "df = df[~df['bin_name'].isin(remove_bins)]\n",
    "df = df.sort_values(by='bin')\n",
    "columns_to_drop = [x for x in df.columns if x in adata.obs.columns]\n",
    "df = df.drop(columns=columns_to_drop)\n",
    "print(f\"{df.shape=}\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6337ffb4-e3cf-42c6-97dc-2d5e5c2e4fda",
   "metadata": {},
   "source": [
    "# Merge the individual core scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b7c6d1d-51a7-443c-a8a4-20754c537d3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs = pd.merge(\n",
    "    adata.obs, df.set_index('bin_name'),\n",
    "    how='left',\n",
    "    left_index=True,\n",
    "    right_index=True,\n",
    ")\n",
    "\n",
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3a54fa5-75c9-41e8-8ca3-d4b0dc326047",
   "metadata": {},
   "source": [
    "# compute the genome wide core score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bab6679a-f598-41a6-af8d-2bd3aed3e46e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # add the principal singular value of the incidence matrix\n",
    "# start_time = time.time()\n",
    "# H = adata.to_df().copy()\n",
    "# print(f\"Raw: {H.shape=}\")\n",
    "# H = H.T.drop_duplicates().T\n",
    "# print(f\"De-duped: {H.shape=}\")\n",
    "# end_time = time.time()\n",
    "# print(f\"Data de-duplication took: {(end_time - start_time)/60:.2f} minutes\")\n",
    "\n",
    "# start_time = time.time()\n",
    "# svd = TruncatedSVD(n_components=1, n_iter=10)\n",
    "# adata.obs['global_singular_vector_1'] = ut.min_max(svd.fit_transform(H))\n",
    "# end_time = time.time()\n",
    "# print(f\"SVD calculation took: {(end_time - start_time)/60:.2f} minutes\")\n",
    "\n",
    "# rna_weights = 1 / (adata.obs.loc[H.index, 'RNA_2'].values + 1)\n",
    "# rna_weights = np.nan_to_num(rna_weights, nan=0) \n",
    "\n",
    "# # hypergraph centralities\n",
    "# hge_functions = {\n",
    "#     'global_hge_logexp_unweighted' : {\n",
    "#         'function' : 'log-exp',\n",
    "#         'weights' : None,\n",
    "#     },\n",
    "#     'global_hge_logexp_RNA_weighted' : {\n",
    "#         'function' : 'log-exp',\n",
    "#         'weights' : rna_weights,\n",
    "#     },\n",
    "# }\n",
    "\n",
    "# hge_centralities = []\n",
    "\n",
    "# for label, d in hge_functions.items():\n",
    "#     start_time = time.time()  # Record start time\n",
    "#     node, edge = central.nonlinear_eigenvector_centrality(\n",
    "#         H,\n",
    "#         function=d['function'],\n",
    "#         node_weights=d['weights'],\n",
    "#     )\n",
    "\n",
    "#     hge_centralities.append(label)\n",
    "#     adata.obs[label] = ut.min_max(node)\n",
    "\n",
    "#     end_time = time.time()  # Record end time\n",
    "#     print(f\"{label} calculation took: {(end_time - start_time)/60:.2f} minutes\")\n",
    "\n",
    "\n",
    "# start_time = time.time()\n",
    "# print('done! saving.....')\n",
    "# outpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/global_core_score/population_mESC_{resolution}_scores.csv\"\n",
    "# obs = adata.obs.copy()\n",
    "# obs = obs.reset_index(drop=False)\n",
    "# obs.to_csv(outpath, index=False)\n",
    "# obs.head()\n",
    "# end_time = time.time()\n",
    "# print(f\"Saving results took: {(end_time - start_time)/60:.2f} minutes\")\n",
    "\n",
    "\n",
    "# \"\"\"\n",
    "# Raw: H.shape=(2431, 2756467)\n",
    "# De-duped: H.shape=(2431, 1861825)\n",
    "# Data de-duplication took: 10.31 minutes\n",
    "# SVD calculation took: 42.49 minutes\n",
    "# global_hge_logexp_unweighted calculation took: 5.39 minutes\n",
    "# global_hge_logexp_RNA_weighted calculation took: 5.75 minutes\n",
    "# done! saving.....\n",
    "# Saving results took: 0.00 minutes\n",
    "# \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16be760d-b11d-4d99-95d6-a1cd4cea66c1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74bbda35-56e8-4877-bcd6-23360b2d229f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35481f11-2115-4019-a2f2-25c01b51e889",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42a40464-2c66-4ee5-be5f-8d39b80fbb07",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1e66eb3-5c0a-45f4-9f1b-4a691202aa33",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6322aced-65fd-4263-899b-907f9b3023b9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8d265bf-d508-4ee2-9147-0073bdec7354",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d3c043b-1cc9-4ff6-a3a3-1e62a4207b03",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "957582a7-2c25-40fa-a3f8-00f7e7800a5a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5ee7556-5b76-422e-a8c6-f2c1d5ea32e8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13ea8422-ffd3-470d-a976-a056079dd575",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31fa10a1-ac40-4d78-95db-ddccc5b012c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b677c01f-28f7-4678-8752-f7d7e2383104",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # add the principal singular value of the incidence matrix\n",
    "# H = adata.to_df().copy()\n",
    "# print(f\"Raw: {H.shape=}\")\n",
    "# H = H.T.drop_duplicates().T\n",
    "# print(f\"De-duped: {H.shape=}\")\n",
    "\n",
    "# svd = TruncatedSVD(n_components=1, n_iter=10)\n",
    "# adata.obs['global_singular_vector_1'] = ut.min_max(svd.fit_transform(H))\n",
    "\n",
    "# # hypergraph centralities\n",
    "# hge_functions = {\n",
    "#     'global_hge_logexp_unweighted' : {\n",
    "#         'function' : 'log-exp',\n",
    "#         'weights' : None,\n",
    "#     },\n",
    "#     'global_hge_logexp_RNA_weighted' : {\n",
    "#         'function' : 'log-exp',\n",
    "#         'weights' : 1 / (adata.obs.loc[H.index, 'RNA_2'].values + 1)\n",
    "#     },\n",
    "# }\n",
    "\n",
    "# hge_centralities = []\n",
    "\n",
    "# for label, d in hge_functions.items():\n",
    "#     start_time = time.time()  # Record start time\n",
    "#     node, edge = central.nonlinear_eigenvector_centrality(\n",
    "#         H,\n",
    "#         function=d['function'],\n",
    "#         node_weights=d['weights'],\n",
    "#     )\n",
    "\n",
    "#     hge_centralities.append(label)\n",
    "#     adata.obs[label] = ut.min_max(node)\n",
    "\n",
    "#     end_time = time.time()  # Record end time\n",
    "#     print(f\"{label} calculation took: {end_time - start_time:.2f} seconds\")\n",
    "\n",
    "\n",
    "# print('done! saving.....')\n",
    "# outpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/global_core_score/population_mESC_{resolution}_scores.csv\"\n",
    "# obs = adata.obs.copy()\n",
    "# obs = obs.reset_index(drop=False)\n",
    "# obs.to_csv(outpath, index=False)\n",
    "# obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b8280ec-8f9e-456c-9415-439e1636adf8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b60fdd7-1d7b-444c-8a01-dfa74fc12461",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c927118-9442-4fc8-8d7b-906215e9c658",
   "metadata": {},
   "outputs": [],
   "source": [
    "scores.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d530c758-19a0-468f-9690-14abdfdee5d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 1.25, 2.5\n",
    "\n",
    "\n",
    "sns.boxplot(\n",
    "    data=scores,\n",
    "    x='core_labels',\n",
    "    y='n_genes',\n",
    "    showfliers=False,\n",
    "    palette=['lightgrey', 'r'],\n",
    "    linecolor='k',\n",
    "    width=0.65,\n",
    ")\n",
    "\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel(\"value\")\n",
    "# plt.title(\"ATAC-seq\")\n",
    "# plt.title(\"H3K27me3\")\n",
    "plt.title(\"n genes\")\n",
    "plt.gca().tick_params(axis='x', rotation=90)\n",
    "\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a486ddb7-3131-45da-9c4c-0f63ced2af89",
   "metadata": {},
   "outputs": [],
   "source": [
    "scores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "140842e8-64e8-45cc-a72c-0a4aa232e188",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e43b539-fd5f-43f2-93e9-414996f01694",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "491ec5a4-112b-42d9-94f4-dafbdddc2d32",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b969993b-0397-4340-913d-4fdbd00f7e7c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "435b8e66-4606-441e-b2ff-85dd15a5eca0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55d1d318-d848-46ec-9cf9-732c303b440a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "432f7009-101f-4017-aab4-ea6aa90abd31",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33acc560-f5a0-4528-bec8-08274bb06027",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d68f7d1a-9cd2-4ece-8457-edb3ce6c4a2a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cb991d7-b48b-4ada-a8cc-7b7d655b9514",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b660ab3a-63d9-4099-8edf-9587fdc232c8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f9995bc-c6a1-4b72-89b9-17481c40044f",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_reads = 10000\n",
    "\n",
    "H = adata.to_df().sample(n_reads, axis=1)\n",
    "print(f\"{H.shape=}\")\n",
    "H = ut.sort_by_lowest_index(H)\n",
    "\n",
    "rows, cols = np.nonzero(H)\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 8, 4\n",
    "\n",
    "plt.scatter(\n",
    "    cols, rows, \n",
    "    marker=\"s\",\n",
    "    s=1,\n",
    "    c='lightgrey',\n",
    "    alpha=0.75,\n",
    ")\n",
    "\n",
    "plt.gca().invert_yaxis()\n",
    "plt.xticks([])\n",
    "\n",
    "sns.despine()\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be10ddd8-2a76-4c07-b0d0-3f54d8c26c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8b8591c-fd33-4181-8491-4cbbc141939f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e22d0f9-d671-4da8-bab0-7d23bf88dc7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "dpath = \"/scratch/indikar_root/indikar1/shared_data/higher_order/core_scores/\"\n",
    "\n",
    "file_list = sorted(glob.glob(f\"{dpath}*_{resolution}_*\"))\n",
    "print(f\"{len(file_list)} files\")\n",
    "\n",
    "read_columns = [\n",
    "    'bin_name',\n",
    "    'ce_singular_vector_1', \n",
    "    'ce_eigenvector_centrality',\n",
    "    'ce_betweenness_centrality',\n",
    "    'ce_pagerank',\n",
    "    'hge_singular_vector_1',\n",
    "    'hge_logexp_unweighted',\n",
    "    'hge_logexp_degree_weighted',\n",
    "    'hge_logexp_RNA_weighted',\n",
    "    'hge_logexp_ATAC_weighted',\n",
    "]\n",
    "\n",
    "df = []\n",
    "\n",
    "for fpath in file_list:\n",
    "    tmp = pd.read_csv(fpath, usecols=read_columns)\n",
    "    df.append(tmp)\n",
    "\n",
    "df = pd.concat(df, ignore_index=True)\n",
    "df = df.set_index('bin_name')\n",
    "print(f\"{df.shape=}\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d569d876-91d8-4d23-8ceb-e7e124c24d50",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs = pd.merge(\n",
    "    adata.obs, \n",
    "    df, \n",
    "    how='left',\n",
    "    left_index=True,\n",
    "    right_index=True,\n",
    "    \n",
    ")\n",
    "adata.obs.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
