{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "246af4d6-affa-4b3e-bda0-d11888c54ae2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "import gget\n",
    "import sklearn.preprocessing \n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from scipy.cluster.hierarchy import linkage, dendrogram\n",
    "from scipy.cluster.hierarchy import fcluster\n",
    "from importlib import reload\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import patches\n",
    "import seaborn as sns\n",
    "import networkx as nx\n",
    "from matplotlib_venn import venn2\n",
    "from matplotlib_venn import venn3\n",
    "from scipy import sparse\n",
    "import xgi\n",
    "\n",
    "import warnings\n",
    "warnings.simplefilter(\"ignore\", category=RuntimeWarning)\n",
    "\n",
    "source_path = os.path.abspath(\"../../source/\")\n",
    "sys.path.append(source_path)\n",
    "import utils as ut\n",
    "import plotting as plt2\n",
    "import hypercore as hc\n",
    "import matrix as matrix\n",
    "import centrality as central"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a380edc6-621b-482c-9546-e2fda9d62fa9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pdf.shape=(97, 5)\n",
      "pdf['gene_name'].nunique()=93\n",
      "\n",
      "godf.shape=(218, 2)\n",
      "godf['gene_name'].nunique()=218\n",
      "\n",
      "exdf.shape=(23244, 17)\n",
      "exdf['gene_name'].nunique()=23225\n",
      "gdf.shape=(56655, 17)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>length</th>\n",
       "      <th>midpoint</th>\n",
       "      <th>is_tf</th>\n",
       "      <th>is_gene</th>\n",
       "      <th>is_pt_gene</th>\n",
       "      <th>is_expressed</th>\n",
       "      <th>expression</th>\n",
       "      <th>expression_log</th>\n",
       "      <th>mESC_panglaoDB_marker</th>\n",
       "      <th>mESC_GO_marker</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000104478</td>\n",
       "      <td>GM38212</td>\n",
       "      <td>havana</td>\n",
       "      <td>TEC</td>\n",
       "      <td>1</td>\n",
       "      <td>108344806</td>\n",
       "      <td>108347562</td>\n",
       "      <td>2756</td>\n",
       "      <td>108346184</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000104385</td>\n",
       "      <td>GM7449</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>6980783</td>\n",
       "      <td>6981446</td>\n",
       "      <td>663</td>\n",
       "      <td>6981114</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000102135</td>\n",
       "      <td>GM37108</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>6986782</td>\n",
       "      <td>6993812</td>\n",
       "      <td>7030</td>\n",
       "      <td>6990297</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.438255</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000103282</td>\n",
       "      <td>GM37275</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>6999982</td>\n",
       "      <td>7000012</td>\n",
       "      <td>30</td>\n",
       "      <td>6999997</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000101097</td>\n",
       "      <td>GM6679</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>108697864</td>\n",
       "      <td>108699733</td>\n",
       "      <td>1869</td>\n",
       "      <td>108698798</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id gene_name gene_source          gene_biotype Chromosome  \\\n",
       "0  ENSMUSG00000104478   GM38212      havana                   TEC          1   \n",
       "1  ENSMUSG00000104385    GM7449      havana  processed_pseudogene          1   \n",
       "2  ENSMUSG00000102135   GM37108      havana  processed_pseudogene          1   \n",
       "3  ENSMUSG00000103282   GM37275      havana  processed_pseudogene          1   \n",
       "4  ENSMUSG00000101097    GM6679      havana  processed_pseudogene          1   \n",
       "\n",
       "       Start        End  length   midpoint  is_tf  is_gene  is_pt_gene  \\\n",
       "0  108344806  108347562    2756  108346184  False     True       False   \n",
       "1    6980783    6981446     663    6981114  False     True       False   \n",
       "2    6986782    6993812    7030    6990297  False     True       False   \n",
       "3    6999982    7000012      30    6999997  False     True       False   \n",
       "4  108697864  108699733    1869  108698798  False     True       False   \n",
       "\n",
       "   is_expressed  expression  expression_log  mESC_panglaoDB_marker  \\\n",
       "0         False        0.00        0.000000                  False   \n",
       "1         False        0.00        0.000000                  False   \n",
       "2          True        0.55        0.438255                  False   \n",
       "3         False        0.00        0.000000                  False   \n",
       "4         False        0.00        0.000000                  False   \n",
       "\n",
       "   mESC_GO_marker  \n",
       "0           False  \n",
       "1           False  \n",
       "2           False  \n",
       "3           False  \n",
       "4           False  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"load pangloadb\"\n",
    "fpath = \"../../resources/mESC_pangloadb.csv\"\n",
    "pdf = pd.read_csv(fpath)\n",
    "print(f\"{pdf.shape=}\")\n",
    "print(f\"{pdf['gene_name'].nunique()=}\")\n",
    "print()\n",
    "\n",
    "\"\"\" load gene ontology annotations \"\"\"\n",
    "fpath = \"../../resources/stem_cell_population_maintenance.csv\"\n",
    "godf = pd.read_csv(fpath)\n",
    "godf['gene_name'] = godf['gene_name'].str.upper()\n",
    "print(f\"{godf.shape=}\")\n",
    "print(f\"{godf['gene_name'].nunique()=}\")\n",
    "print()\n",
    "\n",
    "\"\"\" load gene expression \"\"\"\n",
    "fpath = \"/scratch/indikar_root/indikar1/shared_data/higher_order/rna/expression.parquet\"\n",
    "exdf = pd.read_parquet(fpath)\n",
    "\n",
    "tpm_columns = [x for x in exdf.columns if \"TPM\" in x]\n",
    "exdf['expression_mean'] = exdf[tpm_columns].mean(axis=1)\n",
    "exdf['gene_name'] = exdf['gene_name'].str.upper()\n",
    "express_map = dict(zip(exdf['gene_name'].values, exdf['expression_mean'].values))\n",
    "\n",
    "exdf = exdf[exdf['expression_mean'] > 0]\n",
    "print(f\"{exdf.shape=}\")\n",
    "print(f\"{exdf['gene_name'].nunique()=}\")\n",
    "\n",
    "\"\"\" Gene locations \"\"\"\n",
    "\n",
    "\n",
    "gene_table_path = \"/scratch/indikar_root/indikar1/shared_data/higher_order/reference/gene_table.parquet\"\n",
    "gdf = pd.read_parquet(gene_table_path)\n",
    "\n",
    "gdf['gene_name'] = gdf['gene_name'].str.upper()\n",
    "\n",
    "gdf['is_gene'] = [True] * len(gdf)\n",
    "gdf['is_pt_gene'] = gdf['gene_biotype'] == 'protein_coding'\n",
    "gdf['is_expressed'] = gdf['gene_name'].isin(exdf['gene_name'].values)\n",
    "gdf['expression'] = gdf['gene_name'].map(express_map)\n",
    "gdf['expression'] = gdf['expression'].fillna(0)\n",
    "gdf['expression_log'] = np.log1p(gdf['expression'])\n",
    "gdf['mESC_panglaoDB_marker'] = gdf['gene_name'].isin(pdf['gene_name'].values)\n",
    "gdf['mESC_GO_marker'] = gdf['gene_name'].isin(godf['gene_name'].values)\n",
    "\n",
    "print(f\"{gdf.shape=}\")\n",
    "\n",
    "gdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "33b9e749-c682-42dd-bf05-e6db6403aa5a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>length</th>\n",
       "      <th>midpoint</th>\n",
       "      <th>is_tf</th>\n",
       "      <th>is_gene</th>\n",
       "      <th>is_pt_gene</th>\n",
       "      <th>is_expressed</th>\n",
       "      <th>expression</th>\n",
       "      <th>expression_log</th>\n",
       "      <th>mESC_panglaoDB_marker</th>\n",
       "      <th>mESC_GO_marker</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000104478</td>\n",
       "      <td>GM38212</td>\n",
       "      <td>havana</td>\n",
       "      <td>TEC</td>\n",
       "      <td>1</td>\n",
       "      <td>108344806</td>\n",
       "      <td>108347562</td>\n",
       "      <td>2756</td>\n",
       "      <td>108346184</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000104385</td>\n",
       "      <td>GM7449</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>6980783</td>\n",
       "      <td>6981446</td>\n",
       "      <td>663</td>\n",
       "      <td>6981114</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000102135</td>\n",
       "      <td>GM37108</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>6986782</td>\n",
       "      <td>6993812</td>\n",
       "      <td>7030</td>\n",
       "      <td>6990297</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.438255</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000103282</td>\n",
       "      <td>GM37275</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>6999982</td>\n",
       "      <td>7000012</td>\n",
       "      <td>30</td>\n",
       "      <td>6999997</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000101097</td>\n",
       "      <td>GM6679</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>108697864</td>\n",
       "      <td>108699733</td>\n",
       "      <td>1869</td>\n",
       "      <td>108698798</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id gene_name gene_source          gene_biotype Chromosome  \\\n",
       "0  ENSMUSG00000104478   GM38212      havana                   TEC          1   \n",
       "1  ENSMUSG00000104385    GM7449      havana  processed_pseudogene          1   \n",
       "2  ENSMUSG00000102135   GM37108      havana  processed_pseudogene          1   \n",
       "3  ENSMUSG00000103282   GM37275      havana  processed_pseudogene          1   \n",
       "4  ENSMUSG00000101097    GM6679      havana  processed_pseudogene          1   \n",
       "\n",
       "       Start        End  length   midpoint  is_tf  is_gene  is_pt_gene  \\\n",
       "0  108344806  108347562    2756  108346184  False     True       False   \n",
       "1    6980783    6981446     663    6981114  False     True       False   \n",
       "2    6986782    6993812    7030    6990297  False     True       False   \n",
       "3    6999982    7000012      30    6999997  False     True       False   \n",
       "4  108697864  108699733    1869  108698798  False     True       False   \n",
       "\n",
       "   is_expressed  expression  expression_log  mESC_panglaoDB_marker  \\\n",
       "0         False        0.00        0.000000                  False   \n",
       "1         False        0.00        0.000000                  False   \n",
       "2          True        0.55        0.438255                  False   \n",
       "3         False        0.00        0.000000                  False   \n",
       "4         False        0.00        0.000000                  False   \n",
       "\n",
       "   mESC_GO_marker  \n",
       "0           False  \n",
       "1           False  \n",
       "2           False  \n",
       "3           False  \n",
       "4           False  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outpath = \"../../resources/genes.parquet\"\n",
    "gdf.to_parquet(outpath, index=False)\n",
    "gdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8b12f4f-87ae-4a8c-b1e2-625a18407de3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "higher_order",
   "language": "python",
   "name": "higher_order"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
