{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2b8b5ff9-1d28-4b9e-9380-30f5ee8a28c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os \n",
    "import numpy as np\n",
    "from pyranges import PyRanges\n",
    "import pairtools\n",
    "import pairtools.lib.headerops as phead"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "15042143-0159-4fc3-ab13-e08b7b6a04bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df.shape=(1353139, 24)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>readID</th>\n",
       "      <th>chrom1</th>\n",
       "      <th>pos1</th>\n",
       "      <th>chrom2</th>\n",
       "      <th>pos2</th>\n",
       "      <th>strand1</th>\n",
       "      <th>strand2</th>\n",
       "      <th>pair_type</th>\n",
       "      <th>mapq1</th>\n",
       "      <th>mapq2</th>\n",
       "      <th>...</th>\n",
       "      <th>algn_ref_span1</th>\n",
       "      <th>algn_ref_span2</th>\n",
       "      <th>matched_bp1</th>\n",
       "      <th>matched_bp2</th>\n",
       "      <th>rfrag1</th>\n",
       "      <th>rfrag_start1</th>\n",
       "      <th>rfrag_end1</th>\n",
       "      <th>rfrag2</th>\n",
       "      <th>rfrag_start2</th>\n",
       "      <th>rfrag_end2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>c984758d-b12e-44a0-86fa-f5bb326837f1</td>\n",
       "      <td>!</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>42850159</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>NU</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>243</td>\n",
       "      <td>0</td>\n",
       "      <td>242</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>183807</td>\n",
       "      <td>42849815</td>\n",
       "      <td>42850205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6b11c966-4911-4640-a640-c3b651524b5b</td>\n",
       "      <td>!</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>160219378</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>NU</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>590</td>\n",
       "      <td>0</td>\n",
       "      <td>589</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>760125</td>\n",
       "      <td>160218791</td>\n",
       "      <td>160219386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6b11c966-4911-4640-a640-c3b651524b5b</td>\n",
       "      <td>!</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>160218505</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NU</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>278</td>\n",
       "      <td>0</td>\n",
       "      <td>276</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>760123</td>\n",
       "      <td>160218281</td>\n",
       "      <td>160218511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6b11c966-4911-4640-a640-c3b651524b5b</td>\n",
       "      <td>1</td>\n",
       "      <td>160218789</td>\n",
       "      <td>1</td>\n",
       "      <td>160218505</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>UU</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>...</td>\n",
       "      <td>590</td>\n",
       "      <td>278</td>\n",
       "      <td>589</td>\n",
       "      <td>276</td>\n",
       "      <td>760124</td>\n",
       "      <td>160218511</td>\n",
       "      <td>160218791</td>\n",
       "      <td>760123</td>\n",
       "      <td>160218281</td>\n",
       "      <td>160218511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8f835bd9-c2ce-4dbd-8784-11672532138c</td>\n",
       "      <td>!</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>72720934</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>NU</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>345258</td>\n",
       "      <td>72720876</td>\n",
       "      <td>72720976</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 readID chrom1       pos1 chrom2       pos2  \\\n",
       "0  c984758d-b12e-44a0-86fa-f5bb326837f1      !          0      4   42850159   \n",
       "1  6b11c966-4911-4640-a640-c3b651524b5b      !          0      1  160219378   \n",
       "2  6b11c966-4911-4640-a640-c3b651524b5b      !          0      1  160218505   \n",
       "3  6b11c966-4911-4640-a640-c3b651524b5b      1  160218789      1  160218505   \n",
       "4  8f835bd9-c2ce-4dbd-8784-11672532138c      !          0     11   72720934   \n",
       "\n",
       "  strand1 strand2 pair_type  mapq1  mapq2  ...  algn_ref_span1  \\\n",
       "0       -       +        NU      0      1  ...               0   \n",
       "1       -       +        NU      0     60  ...               0   \n",
       "2       -       -        NU      0     60  ...               0   \n",
       "3       +       -        UU     60     60  ...             590   \n",
       "4       -       +        NU      0     21  ...               0   \n",
       "\n",
       "   algn_ref_span2  matched_bp1  matched_bp2  rfrag1  rfrag_start1  rfrag_end1  \\\n",
       "0             243            0          242      -1             0           0   \n",
       "1             590            0          589      -1             0           0   \n",
       "2             278            0          276      -1             0           0   \n",
       "3             278          589          276  760124     160218511   160218791   \n",
       "4              61            0           61      -1             0           0   \n",
       "\n",
       "   rfrag2  rfrag_start2  rfrag_end2  \n",
       "0  183807      42849815    42850205  \n",
       "1  760125     160218791   160219386  \n",
       "2  760123     160218281   160218511  \n",
       "3  760123     160218281   160218511  \n",
       "4  345258      72720876    72720976  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def load_pairs(fpath, comment_char=\"#\"):\n",
    "    \"\"\"Loads a pairs table from a tab-separated file.\n",
    "\n",
    "    Args:\n",
    "        fpath (str): The file path to the pairs table.\n",
    "        comment_char (str, optional): The character indicating comment lines. \n",
    "                                     Defaults to \"#\".\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: A Pandas DataFrame containing the pairs table data.\n",
    "    \"\"\"\n",
    "    header = phead.get_header(open(fpath))[0][-1]\n",
    "    header = header.replace(\"#columns: \", \"\")\n",
    "    header = header.split(\" \")\n",
    "    \n",
    "    df = pd.read_csv(fpath, \n",
    "                     sep='\\t', \n",
    "                     header=None, \n",
    "                     names=header, \n",
    "                     comment=\"#\")\n",
    "    return df\n",
    "\n",
    "\n",
    "fpath = '/scratch/indikar_root/indikar1/shared_data/single_cell/expanded/pairs/o1b05.GRCm39.expanded.pairs'\n",
    "df = load_pairs(fpath)\n",
    "print(f\"{df.shape=}\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6a56c37a-4ffa-4bc3-ae22-42e6366f9960",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['readID', 'chrom1', 'pos1', 'chrom2', 'pos2', 'strand1', 'strand2',\n",
       "       'pair_type', 'mapq1', 'mapq2', 'read_len1', 'read_len2',\n",
       "       'algn_read_span1', 'algn_read_span2', 'algn_ref_span1',\n",
       "       'algn_ref_span2', 'matched_bp1', 'matched_bp2', 'rfrag1',\n",
       "       'rfrag_start1', 'rfrag_end1', 'rfrag2', 'rfrag_start2', 'rfrag_end2'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8adb73cc-d895-4a14-aa61-754d08150d1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pdf.shape=(190905, 24)\n",
      "pdf.shape=(189740, 24)\n",
      "pdf.shape=(131727, 24)\n",
      "pdf.shape=(6479, 24)\n",
      "pdf.shape=(4207, 24)\n",
      "pdf.shape=(4207, 24)\n"
     ]
    }
   ],
   "source": [
    "def filter_unmapped(df):\n",
    "    \"\"\"A function to filter the unmapped reads \"\"\"\n",
    "    df = df[(df['chrom1'] != \"!\") & (df['chrom2'] != \"!\")].copy()\n",
    "    return df\n",
    "\n",
    "def filter_adjacent(df, tolerance=1):\n",
    "    \"\"\"A function to filter out fragments\n",
    "    adjacent on the reference \"\"\"\n",
    "    return df[df['rfrag1'].sub(df['rfrag1'].shift(1)).abs() > tolerance].copy()\n",
    "\n",
    "def filter_close_contact(df, bp=1000):\n",
    "    \"\"\"A function to filter contacts which are close\"\"\"\n",
    "    return df[~((df['chrom1'] == df['chrom2']) & (df['pos1'].sub(df['pos2']).abs() < bp))].copy()\n",
    "\n",
    "\n",
    "def filter_duplicate_contacts(df):\n",
    "    return (df.assign(rfrag_diff=df['rfrag1'].sub(df['rfrag2']).abs())\n",
    "              .sort_values('rfrag_diff')\n",
    "              .drop_duplicates(subset=['rfrag1', 'rfrag2'], keep='first')\n",
    "              .drop(columns=['rfrag_diff']))\n",
    "\n",
    "\n",
    "def filter_promiscuous_fragments(df, threshold = 10):\n",
    "    \"\"\"Filters out contacts involving promiscuous fragments from a DataFrame.\n",
    "\n",
    "    Args:\n",
    "        df (pd.DataFrame): DataFrame containing contact information with 'rfrag1' and 'rfrag2' columns.\n",
    "        threshold (int, optional): The minimum number of interactions for a fragment to be considered promiscuous. Defaults to 10.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: The filtered DataFrame with contacts involving promiscuous fragments removed.\n",
    "    \"\"\"\n",
    "\n",
    "    fragment_counts = (pd.concat([df[\"rfrag1\"], df[\"rfrag2\"]])\n",
    "                         .value_counts()\n",
    "                         .reset_index(name=\"freq\")\n",
    "                         .rename(columns={\"index\": \"fragment_id\"}))\n",
    "\n",
    "    promiscuous_fragments = fragment_counts[fragment_counts[\"freq\"] > threshold][\"fragment_id\"].tolist()\n",
    "\n",
    "    return df[~(df[\"rfrag1\"].isin(promiscuous_fragments) | df[\"rfrag2\"].isin(promiscuous_fragments))]\n",
    "\n",
    "\n",
    "def filter_isolated_fragments(df, max_distance = 10_000_000):\n",
    "    \"\"\"Filters out contacts involving isolated fragments from a DataFrame.\n",
    "\n",
    "    Args:\n",
    "        df (pd.DataFrame): DataFrame containing contact information with \n",
    "            'rfrag1', 'rfrag2', 'chrom1', 'pos1', 'chrom2', 'pos2' columns.\n",
    "        max_distance (int, optional): The maximum distance to another fragment for a fragment to be considered non-isolated. Defaults to 10,000,000.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: The filtered DataFrame with contacts involving isolated fragments removed.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Create a combined DataFrame with fragment information\n",
    "    fragment_data = pd.concat([\n",
    "        df[['rfrag1', 'chrom1', 'rfrag_start1', 'rfrag_end1']].rename(columns={'rfrag1': 'id', 'chrom1': 'Chromosome', 'rfrag_start1': 'Start', 'rfrag_end1': 'End'}),\n",
    "        df[['rfrag2', 'chrom2', 'rfrag_start2', 'rfrag_end2']].rename(columns={'rfrag2': 'id', 'chrom2': 'Chromosome', 'rfrag_start2': 'Start', 'rfrag_end2': 'End'}),\n",
    "    ]).drop_duplicates()\n",
    "\n",
    "    # Create PyRanges object\n",
    "    frag_ranges = PyRanges(fragment_data)\n",
    "\n",
    "    # Find nearest neighbors\n",
    "    nearest_fragments = frag_ranges.nearest(frag_ranges)\n",
    "\n",
    "    # Identify isolated fragments\n",
    "    isolated_fragments = nearest_fragments.df[nearest_fragments.df[\"Distance\"] > max_distance][\"id\"].tolist()\n",
    "\n",
    "    # Filter out contacts involving isolated fragments\n",
    "    return df[~(df[\"rfrag1\"].isin(isolated_fragments) | df[\"rfrag2\"].isin(isolated_fragments))]\n",
    "\n",
    "\n",
    "pdf = filter_unmapped(df)\n",
    "print(f\"{pdf.shape=}\")\n",
    "\n",
    "pdf = filter_adjacent(pdf)\n",
    "print(f\"{pdf.shape=}\")\n",
    "\n",
    "pdf = filter_close_contact(pdf)\n",
    "print(f\"{pdf.shape=}\")\n",
    "\n",
    "pdf = filter_duplicate_contacts(pdf)\n",
    "print(f\"{pdf.shape=}\")\n",
    "\n",
    "pdf = filter_promiscuous_fragments(pdf)\n",
    "print(f\"{pdf.shape=}\")\n",
    "\n",
    "pdf = filter_isolated_fragments(pdf)\n",
    "print(f\"{pdf.shape=}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9181a47d-7c1c-45b9-9e6d-cf95225616b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4207, 24)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def contact_filter(df):\n",
    "    \"\"\"Filters contacts, returning the filtered DataFrame and a summary of filter results.\n",
    "\n",
    "    Args:\n",
    "        df (pandas.DataFrame): DataFrame with contact information.\n",
    "\n",
    "    Returns:\n",
    "        tuple: A tuple containing:\n",
    "            - pandas.DataFrame: The final filtered DataFrame.\n",
    "            - pandas.DataFrame: A DataFrame summarizing the number of rows removed by each filter.\n",
    "    \"\"\"\n",
    "\n",
    "    df = df.reset_index(drop=True).copy()  # Ensure a copy to avoid SettingWithCopyWarning\n",
    "    summary_data = []\n",
    "\n",
    "    filters = [\n",
    "        (filter_unmapped, \"unmapped\"),\n",
    "        (filter_adjacent, \"adjacent\"),\n",
    "        (filter_close_contact, \"close_contact\"),\n",
    "        (filter_duplicate_contacts, \"duplicate\"),\n",
    "        (filter_promiscuous_fragments, \"promiscuous\"),\n",
    "        (filter_isolated_fragments, \"isolated\"),\n",
    "    ]\n",
    "\n",
    "    for filter_func, filter_name in filters:\n",
    "        original_size = len(df)\n",
    "        df = filter_func(df)  # Apply the filter\n",
    "        filtered_out_count = original_size - len(df) # Calculate rows removed\n",
    "        summary_data.append({\"filter\": filter_name, \"rows_filtered_out\": filtered_out_count}) # Store summary\n",
    "\n",
    "    summary_df = pd.DataFrame(summary_data)  # Create summary DataFrame\n",
    "    return df, summary_df\n",
    "\n",
    "pdf, sum_df = contact_filter(df)\n",
    "pdf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "556262e3-47d4-4ae4-b783-612fc2fe7c9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filter</th>\n",
       "      <th>rows_filtered_out</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>unmapped</td>\n",
       "      <td>1162234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>adjacent</td>\n",
       "      <td>1165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>close_contact</td>\n",
       "      <td>58013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>duplicate</td>\n",
       "      <td>125248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>promiscuous</td>\n",
       "      <td>2272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>isolated</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          filter  rows_filtered_out\n",
       "0       unmapped            1162234\n",
       "1       adjacent               1165\n",
       "2  close_contact              58013\n",
       "3      duplicate             125248\n",
       "4    promiscuous               2272\n",
       "5       isolated                  0"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "da14c097-cfed-4289-9d6b-25f95448d3d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_pairs(df, header, outpath):\n",
    "    \"\"\"A function to write a pairs file \"\"\"\n",
    "    with open(outpath, \"a\") as file:  # Open in append mode\n",
    "        for line in header:\n",
    "            file.write(line + \"\\n\") \n",
    "        \n",
    "    df.to_csv(outpath, mode='a', header=None, sep=\"\\t\", index=False)\n",
    "    \n",
    "    \n",
    "\n",
    "fpath = '/scratch/indikar_root/indikar1/shared_data/single_cell/expanded/pairs/o1b05.GRCm39.expanded.pairs'\n",
    "header = phead.get_header(open(fpath))[0]\n",
    "\n",
    "outpath = \"/home/cstansbu/temp/test.pairs\"\n",
    "\n",
    "write_pairs(df, header, outpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77549ed5-26da-4228-aec2-084a445bc565",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c54a4627-54d4-4477-adb3-10cacd1a8d2c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "bc493179-496f-484d-8231-33cfc86c3ce4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['## pairs format v1.0.0',\n",
       " '#shape: whole matrix',\n",
       " '#genome_assembly: unknown',\n",
       " '#chromsize: 1 195154279',\n",
       " '#chromsize: 2 181755017',\n",
       " '#chromsize: 3 159745316',\n",
       " '#chromsize: 4 156860686',\n",
       " '#chromsize: 5 151758149',\n",
       " '#chromsize: 6 149588044',\n",
       " '#chromsize: 7 144995196',\n",
       " '#chromsize: 8 130127694',\n",
       " '#chromsize: 9 124359700',\n",
       " '#chromsize: 10 130530862',\n",
       " '#chromsize: 11 121973369',\n",
       " '#chromsize: 12 120092757',\n",
       " '#chromsize: 13 120883175',\n",
       " '#chromsize: 14 125139656',\n",
       " '#chromsize: 15 104073951',\n",
       " '#chromsize: 16 98008968',\n",
       " '#chromsize: 17 95294699',\n",
       " '#chromsize: 18 90720763',\n",
       " '#chromsize: 19 61420004',\n",
       " '#chromsize: X 169476592',\n",
       " '#chromsize: Y 91455967',\n",
       " '#chromsize: MT 16299',\n",
       " '#chromsize: JH584299.1 953012',\n",
       " '#chromsize: GL456233.2 559103',\n",
       " '#chromsize: JH584301.1 259875',\n",
       " '#chromsize: GL456211.1 241735',\n",
       " '#chromsize: GL456221.1 206961',\n",
       " '#chromsize: JH584297.1 205776',\n",
       " '#chromsize: JH584296.1 199368',\n",
       " '#chromsize: GL456354.1 195993',\n",
       " '#chromsize: JH584298.1 184189',\n",
       " '#chromsize: JH584300.1 182347',\n",
       " '#chromsize: GL456219.1 175968',\n",
       " '#chromsize: GL456210.1 169725',\n",
       " '#chromsize: JH584303.1 158099',\n",
       " '#chromsize: JH584302.1 155838',\n",
       " '#chromsize: GL456212.1 153618',\n",
       " '#chromsize: JH584304.1 114452',\n",
       " '#chromsize: GL456379.1 72385',\n",
       " '#chromsize: GL456366.1 47073',\n",
       " '#chromsize: GL456367.1 42057',\n",
       " '#chromsize: GL456239.1 40056',\n",
       " '#chromsize: GL456383.1 38659',\n",
       " '#chromsize: GL456385.1 35240',\n",
       " '#chromsize: GL456360.1 31704',\n",
       " '#chromsize: GL456378.1 31602',\n",
       " '#chromsize: MU069435.1 31129',\n",
       " '#chromsize: GL456389.1 28772',\n",
       " '#chromsize: GL456372.1 28664',\n",
       " '#chromsize: GL456370.1 26764',\n",
       " '#chromsize: GL456381.1 25871',\n",
       " '#chromsize: GL456387.1 24685',\n",
       " '#chromsize: GL456390.1 24668',\n",
       " '#chromsize: GL456394.1 24323',\n",
       " '#chromsize: GL456392.1 23629',\n",
       " '#chromsize: GL456382.1 23158',\n",
       " '#chromsize: GL456359.1 22974',\n",
       " '#chromsize: GL456396.1 21240',\n",
       " '#chromsize: GL456368.1 20208',\n",
       " '#chromsize: MU069434.1 8412',\n",
       " '#chromsize: JH584295.1 1976',\n",
       " '#samheader: @HD\\tVN:1.6\\tSO:unsorted\\tGO:query',\n",
       " '#samheader: @SQ\\tSN:1\\tLN:195154279',\n",
       " '#samheader: @SQ\\tSN:2\\tLN:181755017',\n",
       " '#samheader: @SQ\\tSN:3\\tLN:159745316',\n",
       " '#samheader: @SQ\\tSN:4\\tLN:156860686',\n",
       " '#samheader: @SQ\\tSN:5\\tLN:151758149',\n",
       " '#samheader: @SQ\\tSN:6\\tLN:149588044',\n",
       " '#samheader: @SQ\\tSN:7\\tLN:144995196',\n",
       " '#samheader: @SQ\\tSN:8\\tLN:130127694',\n",
       " '#samheader: @SQ\\tSN:9\\tLN:124359700',\n",
       " '#samheader: @SQ\\tSN:10\\tLN:130530862',\n",
       " '#samheader: @SQ\\tSN:11\\tLN:121973369',\n",
       " '#samheader: @SQ\\tSN:12\\tLN:120092757',\n",
       " '#samheader: @SQ\\tSN:13\\tLN:120883175',\n",
       " '#samheader: @SQ\\tSN:14\\tLN:125139656',\n",
       " '#samheader: @SQ\\tSN:15\\tLN:104073951',\n",
       " '#samheader: @SQ\\tSN:16\\tLN:98008968',\n",
       " '#samheader: @SQ\\tSN:17\\tLN:95294699',\n",
       " '#samheader: @SQ\\tSN:18\\tLN:90720763',\n",
       " '#samheader: @SQ\\tSN:19\\tLN:61420004',\n",
       " '#samheader: @SQ\\tSN:X\\tLN:169476592',\n",
       " '#samheader: @SQ\\tSN:Y\\tLN:91455967',\n",
       " '#samheader: @SQ\\tSN:MT\\tLN:16299',\n",
       " '#samheader: @SQ\\tSN:JH584299.1\\tLN:953012',\n",
       " '#samheader: @SQ\\tSN:GL456233.2\\tLN:559103',\n",
       " '#samheader: @SQ\\tSN:JH584301.1\\tLN:259875',\n",
       " '#samheader: @SQ\\tSN:GL456211.1\\tLN:241735',\n",
       " '#samheader: @SQ\\tSN:GL456221.1\\tLN:206961',\n",
       " '#samheader: @SQ\\tSN:JH584297.1\\tLN:205776',\n",
       " '#samheader: @SQ\\tSN:JH584296.1\\tLN:199368',\n",
       " '#samheader: @SQ\\tSN:GL456354.1\\tLN:195993',\n",
       " '#samheader: @SQ\\tSN:JH584298.1\\tLN:184189',\n",
       " '#samheader: @SQ\\tSN:JH584300.1\\tLN:182347',\n",
       " '#samheader: @SQ\\tSN:GL456219.1\\tLN:175968',\n",
       " '#samheader: @SQ\\tSN:GL456210.1\\tLN:169725',\n",
       " '#samheader: @SQ\\tSN:JH584303.1\\tLN:158099',\n",
       " '#samheader: @SQ\\tSN:JH584302.1\\tLN:155838',\n",
       " '#samheader: @SQ\\tSN:GL456212.1\\tLN:153618',\n",
       " '#samheader: @SQ\\tSN:JH584304.1\\tLN:114452',\n",
       " '#samheader: @SQ\\tSN:GL456379.1\\tLN:72385',\n",
       " '#samheader: @SQ\\tSN:GL456366.1\\tLN:47073',\n",
       " '#samheader: @SQ\\tSN:GL456367.1\\tLN:42057',\n",
       " '#samheader: @SQ\\tSN:GL456239.1\\tLN:40056',\n",
       " '#samheader: @SQ\\tSN:GL456383.1\\tLN:38659',\n",
       " '#samheader: @SQ\\tSN:GL456385.1\\tLN:35240',\n",
       " '#samheader: @SQ\\tSN:GL456360.1\\tLN:31704',\n",
       " '#samheader: @SQ\\tSN:GL456378.1\\tLN:31602',\n",
       " '#samheader: @SQ\\tSN:MU069435.1\\tLN:31129',\n",
       " '#samheader: @SQ\\tSN:GL456389.1\\tLN:28772',\n",
       " '#samheader: @SQ\\tSN:GL456372.1\\tLN:28664',\n",
       " '#samheader: @SQ\\tSN:GL456370.1\\tLN:26764',\n",
       " '#samheader: @SQ\\tSN:GL456381.1\\tLN:25871',\n",
       " '#samheader: @SQ\\tSN:GL456387.1\\tLN:24685',\n",
       " '#samheader: @SQ\\tSN:GL456390.1\\tLN:24668',\n",
       " '#samheader: @SQ\\tSN:GL456394.1\\tLN:24323',\n",
       " '#samheader: @SQ\\tSN:GL456392.1\\tLN:23629',\n",
       " '#samheader: @SQ\\tSN:GL456382.1\\tLN:23158',\n",
       " '#samheader: @SQ\\tSN:GL456359.1\\tLN:22974',\n",
       " '#samheader: @SQ\\tSN:GL456396.1\\tLN:21240',\n",
       " '#samheader: @SQ\\tSN:GL456368.1\\tLN:20208',\n",
       " '#samheader: @SQ\\tSN:MU069434.1\\tLN:8412',\n",
       " '#samheader: @SQ\\tSN:JH584295.1\\tLN:1976',\n",
       " '#samheader: @PG\\tPN:minimap2\\tID:minimap2\\tVN:2.28-r1209\\tCL:minimap2 -ay -t 8 -x map-ont /scratch/indikar_root/indikar1/shared_data/single_cell/references/GRCm39.fa -',\n",
       " '#samheader: @PG\\tPN:pore-c-py\\tID:pore-c-py-2\\tVN:2.1.2\\tPP:minimap2\\tCL:/home/cstansbu/miniconda3/envs/pore_c/bin/pore-c-py annotate /scratch/indikar_root/indikar1/shared_data/single_cell/digest/o1b05.GRCm39.bam /scratch/indikar_root/indikar1/shared_data/single_cell/expanded/annotate/o1b05.GRCm39 --force --paired_end --monomers --summary --stdout --threads 8',\n",
       " '#samheader: @PG\\tID:pairtools_parse2\\tPN:pairtools_parse2\\tCL:/home/cstansbu/miniconda3/envs/pairtools/bin/pairtools parse2 --output-stats /scratch/indikar_root/indikar1/shared_data/single_cell/reports/pairtools/expanded/o1b05.GRCm39.stats.txt --chroms-path /scratch/indikar_root/indikar1/shared_data/single_cell/references/GRCm39.chrom.sizes --drop-sam --drop-seq --expand --report-position junction --report-orientation read --readid-transform readID.split(\":\")[0] --add-columns mapq,read_len,algn_read_span,algn_ref_span,matched_bp /scratch/indikar_root/indikar1/shared_data/single_cell/expanded/annotate/o1b05.GRCm39.ns.bam\\tPP:pore-c-py-2\\tVN:1.0.3',\n",
       " '#samheader: @PG\\tID:pairtools_restrict\\tPN:pairtools_restrict\\tCL:/home/cstansbu/miniconda3/envs/pairtools/bin/pairtools restrict -f /scratch/indikar_root/indikar1/shared_data/single_cell/references/GRCm39.fragments.bed -o /scratch/indikar_root/indikar1/shared_data/single_cell/expanded/pairs/o1b05.GRCm39.expanded.pairs\\tPP:pairtools_parse2\\tVN:1.0.3',\n",
       " '#columns: readID chrom1 pos1 chrom2 pos2 strand1 strand2 pair_type mapq1 mapq2 read_len1 read_len2 algn_read_span1 algn_read_span2 algn_ref_span1 algn_ref_span2 matched_bp1 matched_bp2 rfrag1 rfrag_start1 rfrag_end1 rfrag2 rfrag_start2 rfrag_end2']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = '/scratch/indikar_root/indikar1/shared_data/single_cell/expanded/pairs/o1b05.GRCm39.expanded.pairs'\n",
    "phead.get_header(open(fpath))[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10cff185-562e-471b-ad7a-9554e4806aaf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6007ceb8-be9b-4859-9c5b-ecb4753bcafb",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "'readID', \n",
    "'chrom1', \n",
    "'pos1',\n",
    "'chrom2',\n",
    "'pos2', \n",
    "'mapq1',\n",
    "'mapq2', \n",
    "'rfrag1',\n",
    "'rfrag2', \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "376b7e1a-5636-4f69-a79a-3c68f9d99d67",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinf",
   "language": "python",
   "name": "bioinf"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
