{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "34fecaac-c8de-445c-9213-fda019758ea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import time\n",
    "import gget\n",
    "import scipy\n",
    "import matplotlib.patches as patches\n",
    "from scipy.sparse import csr_matrix\n",
    "import anndata as an\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import math\n",
    "from scipy.stats import fisher_exact\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.cm as cm\n",
    "from pycirclize import Circos\n",
    "from scipy.interpolate import splprep, splev\n",
    "import networkx as nx\n",
    "import random\n",
    "from importlib import reload\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.preprocessing import MultiLabelBinarizer\n",
    "import warnings\n",
    "from itertools import combinations\n",
    "import ot\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from matplotlib.colors import ListedColormap\n",
    "\n",
    "import surprise as sup\n",
    "\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "\n",
    "\"\"\"WARNING: no warnings\"\"\"\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# local imports\n",
    "import anndata_utils as anntools\n",
    "\n",
    "source_path = os.path.abspath(\"../source/\")\n",
    "sys.path.append(source_path)\n",
    "import centrality as central\n",
    "import matrix\n",
    "import utils as ut\n",
    "import plotting as plt2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ce52919-2595-4a27-bc8f-49a4de51b27d",
   "metadata": {},
   "source": [
    "# load marker genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "118571dc-f497-4f0c-9012-dff1a02da8aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Gjb1',\n",
       " 'Amotl2',\n",
       " 'Yap1',\n",
       " 'Fbln1',\n",
       " 'Uaca',\n",
       " 'Antxr1',\n",
       " 'Fermt2',\n",
       " 'Serpinh1',\n",
       " 'Pls3',\n",
       " 'Gpx8']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def load_pathway(fpath):\n",
    "    \"\"\"\n",
    "    Loads an Enrichr-like database file into a boolean DataFrame.\n",
    "\n",
    "    Args:\n",
    "        fpath (str): Path to the Enrichr-like database file.\n",
    "\n",
    "    Returns:\n",
    "        pandas.DataFrame: A boolean DataFrame where:\n",
    "            - Index: Genes\n",
    "            - Columns: Pathways\n",
    "            - Values: True if the gene is in the pathway, False otherwise.\n",
    "    \"\"\"\n",
    "\n",
    "    result = []\n",
    "    with open(fpath,  encoding='utf-8') as f:\n",
    "        for line in f:\n",
    "            split_line = [x for x in line.strip().split('\\t') if x]  # Remove empty strings directly\n",
    "\n",
    "            row = {'label': split_line[0]}\n",
    "            for gene in split_line[1:]:\n",
    "                row[gene] = 1\n",
    "\n",
    "            result.append(row)\n",
    "\n",
    "    df = pd.DataFrame(result)\n",
    "    df = df.fillna(0.0).set_index('label').astype(bool).T  # Chained operations for clarity\n",
    "\n",
    "    return df\n",
    "\n",
    "fpath = \"../../ONT-single-cell/resources/PanglaoDB_Augmented_2021.txt\"\n",
    "pdf = load_pathway(fpath)\n",
    "stem_genes = list(pdf[pdf['Embryonic Stem Cells']].index)\n",
    "stem_genes = [x.title() for x in stem_genes]\n",
    "stem_genes[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93e87411-67ce-45d2-8b98-12dc755f1842",
   "metadata": {},
   "source": [
    "# Load scores and expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b1e6c213-9235-44eb-a3b2-b29f81dda029",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scores.shape=(2431, 36)\n",
      "(raw) tdf.shape=(51883, 19)\n",
      "(filtered) tdf.shape=(51883, 19)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id(s)</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "      <th>ens_gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6280</th>\n",
       "      <td>ENSMUSG00000027746.13</td>\n",
       "      <td>ENSMUST00000029309.12,ENSMUST00000122330.1,ENS...</td>\n",
       "      <td>996.08</td>\n",
       "      <td>745.23</td>\n",
       "      <td>619.00</td>\n",
       "      <td>61.51</td>\n",
       "      <td>45.88</td>\n",
       "      <td>618.98</td>\n",
       "      <td>0.15</td>\n",
       "      <td>60.61</td>\n",
       "      <td>47.39</td>\n",
       "      <td>48.766900</td>\n",
       "      <td>72.542000</td>\n",
       "      <td>0.067113</td>\n",
       "      <td>38.272100</td>\n",
       "      <td>56.863700</td>\n",
       "      <td>0.067058</td>\n",
       "      <td>ENSMUSG00000027746</td>\n",
       "      <td>Ufm1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31448</th>\n",
       "      <td>ENSMUSG00000094342.2</td>\n",
       "      <td>ENSMUST00000178144.1,ENSMUST00000186631.1</td>\n",
       "      <td>336.00</td>\n",
       "      <td>123.17</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.841215</td>\n",
       "      <td>0.655416</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.657775</td>\n",
       "      <td>0.655401</td>\n",
       "      <td>ENSMUSG00000094342</td>\n",
       "      <td>Gm17522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20482</th>\n",
       "      <td>ENSMUSG00000074283.5</td>\n",
       "      <td>ENSMUST00000037448.6,ENSMUST00000206362.1,ENSM...</td>\n",
       "      <td>2378.65</td>\n",
       "      <td>2126.89</td>\n",
       "      <td>81.00</td>\n",
       "      <td>2.82</td>\n",
       "      <td>2.10</td>\n",
       "      <td>81.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.75</td>\n",
       "      <td>2.15</td>\n",
       "      <td>2.110160</td>\n",
       "      <td>3.416010</td>\n",
       "      <td>0.081786</td>\n",
       "      <td>1.645350</td>\n",
       "      <td>2.667460</td>\n",
       "      <td>0.081730</td>\n",
       "      <td>ENSMUSG00000074283</td>\n",
       "      <td>Zfp109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26865</th>\n",
       "      <td>ENSMUSG00000085886.1</td>\n",
       "      <td>ENSMUST00000148318.1</td>\n",
       "      <td>2536.00</td>\n",
       "      <td>2284.24</td>\n",
       "      <td>4.00</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.10</td>\n",
       "      <td>4.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.038109</td>\n",
       "      <td>0.291886</td>\n",
       "      <td>0.302365</td>\n",
       "      <td>0.029778</td>\n",
       "      <td>0.228216</td>\n",
       "      <td>0.302576</td>\n",
       "      <td>ENSMUSG00000085886</td>\n",
       "      <td>D030047H15Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29988</th>\n",
       "      <td>ENSMUSG00000091277.2</td>\n",
       "      <td>ENSMUST00000169406.2</td>\n",
       "      <td>1184.00</td>\n",
       "      <td>932.24</td>\n",
       "      <td>6.81</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.40</td>\n",
       "      <td>7.54</td>\n",
       "      <td>2.39</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.166885</td>\n",
       "      <td>1.219090</td>\n",
       "      <td>0.298526</td>\n",
       "      <td>0.130474</td>\n",
       "      <td>0.953441</td>\n",
       "      <td>0.298568</td>\n",
       "      <td>ENSMUSG00000091277</td>\n",
       "      <td>Gm1818</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene_id  \\\n",
       "6280   ENSMUSG00000027746.13   \n",
       "31448   ENSMUSG00000094342.2   \n",
       "20482   ENSMUSG00000074283.5   \n",
       "26865   ENSMUSG00000085886.1   \n",
       "29988   ENSMUSG00000091277.2   \n",
       "\n",
       "                                        transcript_id(s)   length  \\\n",
       "6280   ENSMUST00000029309.12,ENSMUST00000122330.1,ENS...   996.08   \n",
       "31448          ENSMUST00000178144.1,ENSMUST00000186631.1   336.00   \n",
       "20482  ENSMUST00000037448.6,ENSMUST00000206362.1,ENSM...  2378.65   \n",
       "26865                               ENSMUST00000148318.1  2536.00   \n",
       "29988                               ENSMUST00000169406.2  1184.00   \n",
       "\n",
       "       effective_length  expected_count    TPM   FPKM  posterior_mean_count  \\\n",
       "6280             745.23          619.00  61.51  45.88                618.98   \n",
       "31448            123.17            0.00   0.00   0.00                  0.00   \n",
       "20482           2126.89           81.00   2.82   2.10                 81.00   \n",
       "26865           2284.24            4.00   0.13   0.10                  4.00   \n",
       "29988            932.24            6.81   0.54   0.40                  7.54   \n",
       "\n",
       "       posterior_standard_deviation_of_count  pme_TPM  pme_FPKM  \\\n",
       "6280                                    0.15    60.61     47.39   \n",
       "31448                                   0.00     0.28      0.22   \n",
       "20482                                   0.00     2.75      2.15   \n",
       "26865                                   0.00     0.15      0.12   \n",
       "29988                                   2.39     0.64      0.50   \n",
       "\n",
       "       TPM_ci_lower_bound  TPM_ci_upper_bound  \\\n",
       "6280            48.766900           72.542000   \n",
       "31448            0.000003            0.841215   \n",
       "20482            2.110160            3.416010   \n",
       "26865            0.038109            0.291886   \n",
       "29988            0.166885            1.219090   \n",
       "\n",
       "       TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \\\n",
       "6280                                0.067113            38.272100   \n",
       "31448                               0.655416             0.000003   \n",
       "20482                               0.081786             1.645350   \n",
       "26865                               0.302365             0.029778   \n",
       "29988                               0.298526             0.130474   \n",
       "\n",
       "       FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation  \\\n",
       "6280             56.863700                                0.067058   \n",
       "31448             0.657775                                0.655401   \n",
       "20482             2.667460                                0.081730   \n",
       "26865             0.228216                                0.302576   \n",
       "29988             0.953441                                0.298568   \n",
       "\n",
       "              ens_gene_id      gene_name  \n",
       "6280   ENSMUSG00000027746           Ufm1  \n",
       "31448  ENSMUSG00000094342        Gm17522  \n",
       "20482  ENSMUSG00000074283         Zfp109  \n",
       "26865  ENSMUSG00000085886  D030047H15Rik  \n",
       "29988  ENSMUSG00000091277         Gm1818  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\" LOAD centrality \"\"\"\n",
    "resolution = 1000000\n",
    "fpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/global_core_score/population_mESC_{resolution}_scores.csv\"\n",
    "scores = pd.read_csv(fpath)\n",
    "print(f\"{scores.shape=}\")\n",
    "\n",
    "\"\"\" CREATE the bin map for global sorting \"\"\"\n",
    "bin_map = dict(zip(scores['bin_name'].values, scores['bin'].values))\n",
    "\n",
    "\"\"\" LOAD expression \"\"\"\n",
    "fpath = \"/scratch/indikar_root/indikar1/shared_data/higher_order/expression_table/rna_table.parquet\"\n",
    "tdf = pd.read_parquet(fpath)\n",
    "print(f\"(raw) {tdf.shape=}\")\n",
    "print(f\"(filtered) {tdf.shape=}\")\n",
    "\n",
    "expression_map = dict(zip(tdf['gene_name'].values, tdf['TPM'].values))\n",
    "\n",
    "tdf.sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "207099c3-e0dc-4409-99ac-2b57c06cfcda",
   "metadata": {},
   "source": [
    "# Load hyperedges from population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bbeba526-441e-4038-a161-c58e05965861",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "is_core\n",
      "False    3704772\n",
      "True     1234495\n",
      "Name: count, dtype: int64\n",
      "df.shape=(4939267, 19)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>read_name</th>\n",
       "      <th>bin_name</th>\n",
       "      <th>is_tf</th>\n",
       "      <th>is_marker</th>\n",
       "      <th>ATACSeq_1</th>\n",
       "      <th>CTCF</th>\n",
       "      <th>H3K27ac</th>\n",
       "      <th>H3K27me3</th>\n",
       "      <th>RNA_5</th>\n",
       "      <th>core_score</th>\n",
       "      <th>is_core</th>\n",
       "      <th>chrom</th>\n",
       "      <th>bin</th>\n",
       "      <th>expression</th>\n",
       "      <th>order</th>\n",
       "      <th>degree</th>\n",
       "      <th>is_pt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ulk4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3891ee6d-53d1-4ee0-ba2f-3d22291d4493</td>\n",
       "      <td>chr9:121</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.826484</td>\n",
       "      <td>1.149226</td>\n",
       "      <td>1.349552</td>\n",
       "      <td>0.866066</td>\n",
       "      <td>0.573875</td>\n",
       "      <td>0.254849</td>\n",
       "      <td>False</td>\n",
       "      <td>chr9</td>\n",
       "      <td>1394.0</td>\n",
       "      <td>4.12</td>\n",
       "      <td>2</td>\n",
       "      <td>1236</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Smarca2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>66953ddf-e76d-4cdf-aaf8-be028a2d7b04</td>\n",
       "      <td>chr19:26</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.497386</td>\n",
       "      <td>0.547185</td>\n",
       "      <td>0.336787</td>\n",
       "      <td>0.839273</td>\n",
       "      <td>0.141617</td>\n",
       "      <td>0.157878</td>\n",
       "      <td>False</td>\n",
       "      <td>chr19</td>\n",
       "      <td>2436.0</td>\n",
       "      <td>4.55</td>\n",
       "      <td>11</td>\n",
       "      <td>609</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Dlgap3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ad5b2240-893f-4ed0-a157-c2be66d8d754</td>\n",
       "      <td>chr4:127</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.754788</td>\n",
       "      <td>1.027046</td>\n",
       "      <td>1.577616</td>\n",
       "      <td>0.839461</td>\n",
       "      <td>0.906364</td>\n",
       "      <td>0.491722</td>\n",
       "      <td>True</td>\n",
       "      <td>chr4</td>\n",
       "      <td>665.0</td>\n",
       "      <td>87.94</td>\n",
       "      <td>5</td>\n",
       "      <td>262</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gm56531</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>3f354c45-5e48-4f6d-8c7e-05369432b344</td>\n",
       "      <td>chr12:8</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.690311</td>\n",
       "      <td>0.851758</td>\n",
       "      <td>0.774613</td>\n",
       "      <td>0.833768</td>\n",
       "      <td>0.617768</td>\n",
       "      <td>0.310082</td>\n",
       "      <td>False</td>\n",
       "      <td>chr12</td>\n",
       "      <td>1659.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>155</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Twsg1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>d4626feb-16a2-4aac-8145-53e89b60bf7c</td>\n",
       "      <td>chr17:66</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.767515</td>\n",
       "      <td>0.921479</td>\n",
       "      <td>1.337415</td>\n",
       "      <td>0.571748</td>\n",
       "      <td>0.508463</td>\n",
       "      <td>0.289030</td>\n",
       "      <td>False</td>\n",
       "      <td>chr17</td>\n",
       "      <td>2289.0</td>\n",
       "      <td>23.16</td>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name    gene_biotype                             read_name  bin_name  \\\n",
       "0      Ulk4  protein_coding  3891ee6d-53d1-4ee0-ba2f-3d22291d4493  chr9:121   \n",
       "1   Smarca2  protein_coding  66953ddf-e76d-4cdf-aaf8-be028a2d7b04  chr19:26   \n",
       "2    Dlgap3  protein_coding  ad5b2240-893f-4ed0-a157-c2be66d8d754  chr4:127   \n",
       "3   Gm56531          lncRNA  3f354c45-5e48-4f6d-8c7e-05369432b344   chr12:8   \n",
       "4     Twsg1  protein_coding  d4626feb-16a2-4aac-8145-53e89b60bf7c  chr17:66   \n",
       "\n",
       "   is_tf  is_marker  ATACSeq_1      CTCF   H3K27ac  H3K27me3     RNA_5  \\\n",
       "0  False      False   0.826484  1.149226  1.349552  0.866066  0.573875   \n",
       "1  False      False   0.497386  0.547185  0.336787  0.839273  0.141617   \n",
       "2  False      False   0.754788  1.027046  1.577616  0.839461  0.906364   \n",
       "3  False      False   0.690311  0.851758  0.774613  0.833768  0.617768   \n",
       "4  False      False   0.767515  0.921479  1.337415  0.571748  0.508463   \n",
       "\n",
       "   core_score  is_core  chrom     bin  expression  order  degree  is_pt  \n",
       "0    0.254849    False   chr9  1394.0        4.12      2    1236   True  \n",
       "1    0.157878    False  chr19  2436.0        4.55     11     609   True  \n",
       "2    0.491722     True   chr4   665.0       87.94      5     262   True  \n",
       "3    0.310082    False  chr12  1659.0         NaN      1     155  False  \n",
       "4    0.289030    False  chr17  2289.0       23.16      1     140   True  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\" LOAD the TF list \"\"\"\n",
    "fpath = \"/nfs/turbo/umms-indikar/shared/projects/twin_cell/data/b_matrix/SCENIC/scenic/tf_lists/allTFs_mm.txt\"\n",
    "tf_list = [x.strip() for x in open(fpath)]\n",
    "\n",
    "\"\"\" LOAD the genes \"\"\"\n",
    "fpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/lightweight/population_mESC_{resolution}_gdf.parquet\"\n",
    "df = pd.read_parquet(fpath)\n",
    "df['is_tf'] = df['gene_name'].isin(tf_list)\n",
    "df['is_marker'] = df['gene_name'].isin(stem_genes)\n",
    "\n",
    "\"\"\" MERGE the core scores \"\"\"\n",
    "merge_columns = [\n",
    "    'bin_name',\n",
    "    'ATACSeq_1',\n",
    "    'CTCF',\n",
    "    'H3K27ac', \n",
    "    'H3K27me3',\n",
    "    'RNA_5',\n",
    "    'global_hge_logexp_RNA_weighted'\n",
    "]\n",
    "\n",
    "df = pd.merge(\n",
    "    df, scores[merge_columns].fillna(0.0), how=\"left\",\n",
    ")\n",
    "\n",
    "\"\"\" DEFINE the core \"\"\"\n",
    "score_column = 'global_hge_logexp_RNA_weighted'\n",
    "core_threshold_quantile = 0.75\n",
    "threshold = np.quantile(df[score_column].fillna(0.0), core_threshold_quantile)\n",
    "df['is_core'] = df[score_column].fillna(0.0) > threshold\n",
    "df = df.rename(columns={\n",
    "    'global_hge_logexp_RNA_weighted' : 'core_score',\n",
    "})\n",
    "\n",
    "print(df['is_core'].value_counts())\n",
    "\n",
    "\"\"\"ADD some convience columns \"\"\"\n",
    "df['chrom'] = df['bin_name'].str.split(\":\").str[0]\n",
    "df['bin'] = df['bin_name'].map(bin_map)\n",
    "df['expression'] = df['gene_name'].map(expression_map)\n",
    "df['order'] = df.groupby('read_name')['gene_name'].transform('nunique')\n",
    "df['degree'] = df.groupby('gene_name')['read_name'].transform('nunique')\n",
    "df['is_pt'] = (df['gene_biotype'] == 'protein_coding')\n",
    "\n",
    "print(f\"{df.shape=}\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b9c4a64-1675-41f2-9285-5b84391da777",
   "metadata": {},
   "source": [
    "# Save the reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cb8135f4-a974-4fb3-82db-022b893bcdbd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame saved to: /scratch/indikar_root/indikar1/shared_data/higher_order/transcription_clusters/read_table_1000000.parquet\n",
      "Total time taken: 5.55 seconds\n",
      "\n",
      "First 5 rows of the DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>read_name</th>\n",
       "      <th>bin_name</th>\n",
       "      <th>is_tf</th>\n",
       "      <th>is_marker</th>\n",
       "      <th>ATACSeq_1</th>\n",
       "      <th>CTCF</th>\n",
       "      <th>H3K27ac</th>\n",
       "      <th>H3K27me3</th>\n",
       "      <th>RNA_5</th>\n",
       "      <th>core_score</th>\n",
       "      <th>is_core</th>\n",
       "      <th>chrom</th>\n",
       "      <th>bin</th>\n",
       "      <th>expression</th>\n",
       "      <th>order</th>\n",
       "      <th>degree</th>\n",
       "      <th>is_pt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ulk4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3891ee6d-53d1-4ee0-ba2f-3d22291d4493</td>\n",
       "      <td>chr9:121</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.826484</td>\n",
       "      <td>1.149226</td>\n",
       "      <td>1.349552</td>\n",
       "      <td>0.866066</td>\n",
       "      <td>0.573875</td>\n",
       "      <td>0.254849</td>\n",
       "      <td>False</td>\n",
       "      <td>chr9</td>\n",
       "      <td>1394.0</td>\n",
       "      <td>4.12</td>\n",
       "      <td>2</td>\n",
       "      <td>1236</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Smarca2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>66953ddf-e76d-4cdf-aaf8-be028a2d7b04</td>\n",
       "      <td>chr19:26</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.497386</td>\n",
       "      <td>0.547185</td>\n",
       "      <td>0.336787</td>\n",
       "      <td>0.839273</td>\n",
       "      <td>0.141617</td>\n",
       "      <td>0.157878</td>\n",
       "      <td>False</td>\n",
       "      <td>chr19</td>\n",
       "      <td>2436.0</td>\n",
       "      <td>4.55</td>\n",
       "      <td>11</td>\n",
       "      <td>609</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Dlgap3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ad5b2240-893f-4ed0-a157-c2be66d8d754</td>\n",
       "      <td>chr4:127</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.754788</td>\n",
       "      <td>1.027046</td>\n",
       "      <td>1.577616</td>\n",
       "      <td>0.839461</td>\n",
       "      <td>0.906364</td>\n",
       "      <td>0.491722</td>\n",
       "      <td>True</td>\n",
       "      <td>chr4</td>\n",
       "      <td>665.0</td>\n",
       "      <td>87.94</td>\n",
       "      <td>5</td>\n",
       "      <td>262</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gm56531</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>3f354c45-5e48-4f6d-8c7e-05369432b344</td>\n",
       "      <td>chr12:8</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.690311</td>\n",
       "      <td>0.851758</td>\n",
       "      <td>0.774613</td>\n",
       "      <td>0.833768</td>\n",
       "      <td>0.617768</td>\n",
       "      <td>0.310082</td>\n",
       "      <td>False</td>\n",
       "      <td>chr12</td>\n",
       "      <td>1659.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>155</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Twsg1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>d4626feb-16a2-4aac-8145-53e89b60bf7c</td>\n",
       "      <td>chr17:66</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.767515</td>\n",
       "      <td>0.921479</td>\n",
       "      <td>1.337415</td>\n",
       "      <td>0.571748</td>\n",
       "      <td>0.508463</td>\n",
       "      <td>0.289030</td>\n",
       "      <td>False</td>\n",
       "      <td>chr17</td>\n",
       "      <td>2289.0</td>\n",
       "      <td>23.16</td>\n",
       "      <td>1</td>\n",
       "      <td>140</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name    gene_biotype                             read_name  bin_name  \\\n",
       "0      Ulk4  protein_coding  3891ee6d-53d1-4ee0-ba2f-3d22291d4493  chr9:121   \n",
       "1   Smarca2  protein_coding  66953ddf-e76d-4cdf-aaf8-be028a2d7b04  chr19:26   \n",
       "2    Dlgap3  protein_coding  ad5b2240-893f-4ed0-a157-c2be66d8d754  chr4:127   \n",
       "3   Gm56531          lncRNA  3f354c45-5e48-4f6d-8c7e-05369432b344   chr12:8   \n",
       "4     Twsg1  protein_coding  d4626feb-16a2-4aac-8145-53e89b60bf7c  chr17:66   \n",
       "\n",
       "   is_tf  is_marker  ATACSeq_1      CTCF   H3K27ac  H3K27me3     RNA_5  \\\n",
       "0  False      False   0.826484  1.149226  1.349552  0.866066  0.573875   \n",
       "1  False      False   0.497386  0.547185  0.336787  0.839273  0.141617   \n",
       "2  False      False   0.754788  1.027046  1.577616  0.839461  0.906364   \n",
       "3  False      False   0.690311  0.851758  0.774613  0.833768  0.617768   \n",
       "4  False      False   0.767515  0.921479  1.337415  0.571748  0.508463   \n",
       "\n",
       "   core_score  is_core  chrom     bin  expression  order  degree  is_pt  \n",
       "0    0.254849    False   chr9  1394.0        4.12      2    1236   True  \n",
       "1    0.157878    False  chr19  2436.0        4.55     11     609   True  \n",
       "2    0.491722     True   chr4   665.0       87.94      5     262   True  \n",
       "3    0.310082    False  chr12  1659.0         NaN      1     155  False  \n",
       "4    0.289030    False  chr17  2289.0       23.16      1     140   True  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "\n",
    "outpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/transcription_clusters/read_table_{resolution}.parquet\"\n",
    "\n",
    "df.to_parquet(outpath, index=False)\n",
    "\n",
    "end_time = time.time()\n",
    "total_time = end_time - start_time\n",
    "\n",
    "print(f\"DataFrame saved to: {outpath}\")\n",
    "print(f\"Total time taken: {total_time:.2f} seconds\")\n",
    "\n",
    "print(\"\\nFirst 5 rows of the DataFrame:\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e43abb93-dee3-4f91-b6aa-08195faaf8f8",
   "metadata": {},
   "source": [
    "# make gene and read references"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "41e94d4f-7066-41c5-9d55-c975988ba286",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nodes.shape=(48681, 17)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>bin_name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>bin</th>\n",
       "      <th>is_pt</th>\n",
       "      <th>is_tf</th>\n",
       "      <th>is_marker</th>\n",
       "      <th>is_core</th>\n",
       "      <th>core_score</th>\n",
       "      <th>ATACSeq_1</th>\n",
       "      <th>CTCF</th>\n",
       "      <th>H3K27ac</th>\n",
       "      <th>H3K27me3</th>\n",
       "      <th>RNA_5</th>\n",
       "      <th>expression</th>\n",
       "      <th>degree</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ulk4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr9:121</td>\n",
       "      <td>chr9</td>\n",
       "      <td>1394.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.254849</td>\n",
       "      <td>0.826484</td>\n",
       "      <td>1.149226</td>\n",
       "      <td>1.349552</td>\n",
       "      <td>0.866066</td>\n",
       "      <td>0.573875</td>\n",
       "      <td>4.12</td>\n",
       "      <td>1236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Smarca2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr19:26</td>\n",
       "      <td>chr19</td>\n",
       "      <td>2436.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.157878</td>\n",
       "      <td>0.497386</td>\n",
       "      <td>0.547185</td>\n",
       "      <td>0.336787</td>\n",
       "      <td>0.839273</td>\n",
       "      <td>0.141617</td>\n",
       "      <td>4.55</td>\n",
       "      <td>609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Dlgap3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr4:127</td>\n",
       "      <td>chr4</td>\n",
       "      <td>665.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.491722</td>\n",
       "      <td>0.754788</td>\n",
       "      <td>1.027046</td>\n",
       "      <td>1.577616</td>\n",
       "      <td>0.839461</td>\n",
       "      <td>0.906364</td>\n",
       "      <td>87.94</td>\n",
       "      <td>262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gm56531</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr12:8</td>\n",
       "      <td>chr12</td>\n",
       "      <td>1659.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.310082</td>\n",
       "      <td>0.690311</td>\n",
       "      <td>0.851758</td>\n",
       "      <td>0.774613</td>\n",
       "      <td>0.833768</td>\n",
       "      <td>0.617768</td>\n",
       "      <td>NaN</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Twsg1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr17:66</td>\n",
       "      <td>chr17</td>\n",
       "      <td>2289.0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.289030</td>\n",
       "      <td>0.767515</td>\n",
       "      <td>0.921479</td>\n",
       "      <td>1.337415</td>\n",
       "      <td>0.571748</td>\n",
       "      <td>0.508463</td>\n",
       "      <td>23.16</td>\n",
       "      <td>140</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name    gene_biotype  bin_name  chrom     bin  is_pt  is_tf  is_marker  \\\n",
       "0      Ulk4  protein_coding  chr9:121   chr9  1394.0   True  False      False   \n",
       "1   Smarca2  protein_coding  chr19:26  chr19  2436.0   True  False      False   \n",
       "2    Dlgap3  protein_coding  chr4:127   chr4   665.0   True  False      False   \n",
       "3   Gm56531          lncRNA   chr12:8  chr12  1659.0  False  False      False   \n",
       "4     Twsg1  protein_coding  chr17:66  chr17  2289.0   True  False      False   \n",
       "\n",
       "   is_core  core_score  ATACSeq_1      CTCF   H3K27ac  H3K27me3     RNA_5  \\\n",
       "0    False    0.254849   0.826484  1.149226  1.349552  0.866066  0.573875   \n",
       "1    False    0.157878   0.497386  0.547185  0.336787  0.839273  0.141617   \n",
       "2     True    0.491722   0.754788  1.027046  1.577616  0.839461  0.906364   \n",
       "3    False    0.310082   0.690311  0.851758  0.774613  0.833768  0.617768   \n",
       "4    False    0.289030   0.767515  0.921479  1.337415  0.571748  0.508463   \n",
       "\n",
       "   expression  degree  \n",
       "0        4.12    1236  \n",
       "1        4.55     609  \n",
       "2       87.94     262  \n",
       "3         NaN     155  \n",
       "4       23.16     140  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = [\n",
    "    'gene_name', \n",
    "    'gene_biotype', \n",
    "    'bin_name', \n",
    "    'chrom',\n",
    "    'bin',\n",
    "    'is_pt',\n",
    "    'is_tf',\n",
    "    'is_marker',\n",
    "    'is_core', \n",
    "    'core_score',\n",
    "    'ATACSeq_1', \n",
    "    'CTCF', \n",
    "    'H3K27ac', \n",
    "    'H3K27me3',\n",
    "    'RNA_5',\n",
    "    'expression', \n",
    "    'degree'\n",
    "]\n",
    "nodes = df[columns].copy()\n",
    "nodes = nodes.drop_duplicates()\n",
    "print(f\"{nodes.shape=}\")\n",
    "\n",
    "outpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/transcription_clusters/nodes_{resolution}.parquet\"\n",
    "nodes.to_parquet(outpath, index=False)\n",
    "nodes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6f298d8c-05b3-4ffa-b1ae-650e1943368c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "edges.shape=(2200583, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_name</th>\n",
       "      <th>order</th>\n",
       "      <th>n_chroms</th>\n",
       "      <th>n_tf</th>\n",
       "      <th>n_pt</th>\n",
       "      <th>mean_expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00000202-49cf-47b2-83bf-5eb3f6d98373</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>10.170000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00000b61-7794-4b29-9f89-2b74e7bbce3e</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>61.916667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>00001034-f0ab-41c2-885a-da17c2b836ae</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.295000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0000131a-4f27-4dc5-839d-09720b024db9</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>12.840000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>00001afb-6b62-4d01-be02-9509485330ab</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.040000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              read_name  order  n_chroms  n_tf  n_pt  \\\n",
       "0  00000202-49cf-47b2-83bf-5eb3f6d98373      1         1     0     1   \n",
       "1  00000b61-7794-4b29-9f89-2b74e7bbce3e      3         3     0     3   \n",
       "2  00001034-f0ab-41c2-885a-da17c2b836ae      2         2     0     2   \n",
       "3  0000131a-4f27-4dc5-839d-09720b024db9      1         1     0     1   \n",
       "4  00001afb-6b62-4d01-be02-9509485330ab      1         1     0     1   \n",
       "\n",
       "   mean_expression  \n",
       "0        10.170000  \n",
       "1        61.916667  \n",
       "2         0.295000  \n",
       "3        12.840000  \n",
       "4         0.040000  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges = df.groupby('read_name').agg(\n",
    "    order = ('gene_name', 'nunique'),\n",
    "    n_chroms = ('chrom', 'nunique'),\n",
    "    n_tf = ('is_tf', 'sum'),\n",
    "    n_pt = ('is_pt', 'sum'),\n",
    "    mean_expression = ('expression', 'mean'),\n",
    ").reset_index()\n",
    "\n",
    "print(f\"{edges.shape=}\")\n",
    "\n",
    "outpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/transcription_clusters/edges_{resolution}.parquet\"\n",
    "edges.to_parquet(outpath, index=False)\n",
    "edges.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea8d0c19-af93-4e4c-a240-fbc4f5524e46",
   "metadata": {},
   "source": [
    "# Make and store incidence "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "db3a3b5a-32b6-431b-be71-5f423b009488",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H.shape=(21278, 1089002)\n",
      "type(H)=<class 'pandas.core.frame.DataFrame'>\n",
      "DataFrame size: 35.84 MB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3891ee6d-53d1-4ee0-ba2f-3d22291d4493</th>\n",
       "      <th>66953ddf-e76d-4cdf-aaf8-be028a2d7b04</th>\n",
       "      <th>ad5b2240-893f-4ed0-a157-c2be66d8d754</th>\n",
       "      <th>3b213af3-2882-4f23-887e-ecfc864e0f17</th>\n",
       "      <th>5f9cdb29-7ddc-4d9a-93ee-e82edb44a08d</th>\n",
       "      <th>52a1f5d7-c3fc-4cce-bab6-06f8cda9f4bd</th>\n",
       "      <th>d7dbae9b-5deb-4a0f-8f97-eb5101d949cd</th>\n",
       "      <th>4d19c76d-2583-4308-b495-77f34035ac87</th>\n",
       "      <th>e6f27aaf-9cfa-4bf0-b3e4-c81bc946d811</th>\n",
       "      <th>7b85fb37-32dc-4a77-ad2c-26b8c6e57ec3</th>\n",
       "      <th>...</th>\n",
       "      <th>6f6855ef-e494-4a8a-bd4e-7bb05c916950</th>\n",
       "      <th>401c8d32-d3bb-4c71-816e-3581a27f3b9e</th>\n",
       "      <th>0c1c9019-79e9-4d78-84f8-ab977c1b976a</th>\n",
       "      <th>ea773a8a-e760-4bc3-9f3e-bb42e980b45c</th>\n",
       "      <th>3a1950eb-15c2-4b46-b1f4-cf988afc9af8</th>\n",
       "      <th>29884a9e-228d-456a-9869-c663e2b4cb86</th>\n",
       "      <th>dfbc441f-bff3-4972-8ac6-fa69454347a3</th>\n",
       "      <th>a89b8cf7-8579-4bd3-ad1f-a6e4295ff8cc</th>\n",
       "      <th>e13378e4-38f6-4bac-92e8-d849bbd1c764</th>\n",
       "      <th>ea4352dd-04b6-46ae-aae3-0f1af9ec1b1a</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ulk4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Smarca2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dlgap3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tmem267</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Atg4a</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 1089002 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         3891ee6d-53d1-4ee0-ba2f-3d22291d4493  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         66953ddf-e76d-4cdf-aaf8-be028a2d7b04  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         ad5b2240-893f-4ed0-a157-c2be66d8d754  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         3b213af3-2882-4f23-887e-ecfc864e0f17  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         5f9cdb29-7ddc-4d9a-93ee-e82edb44a08d  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         52a1f5d7-c3fc-4cce-bab6-06f8cda9f4bd  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         d7dbae9b-5deb-4a0f-8f97-eb5101d949cd  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         4d19c76d-2583-4308-b495-77f34035ac87  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         e6f27aaf-9cfa-4bf0-b3e4-c81bc946d811  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         7b85fb37-32dc-4a77-ad2c-26b8c6e57ec3  ...  \\\n",
       "Ulk4                                        0  ...   \n",
       "Smarca2                                     0  ...   \n",
       "Dlgap3                                      0  ...   \n",
       "Tmem267                                     0  ...   \n",
       "Atg4a                                       0  ...   \n",
       "\n",
       "         6f6855ef-e494-4a8a-bd4e-7bb05c916950  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         401c8d32-d3bb-4c71-816e-3581a27f3b9e  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         0c1c9019-79e9-4d78-84f8-ab977c1b976a  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         ea773a8a-e760-4bc3-9f3e-bb42e980b45c  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         3a1950eb-15c2-4b46-b1f4-cf988afc9af8  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         29884a9e-228d-456a-9869-c663e2b4cb86  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         dfbc441f-bff3-4972-8ac6-fa69454347a3  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         a89b8cf7-8579-4bd3-ad1f-a6e4295ff8cc  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         e13378e4-38f6-4bac-92e8-d849bbd1c764  \\\n",
       "Ulk4                                        0   \n",
       "Smarca2                                     0   \n",
       "Dlgap3                                      0   \n",
       "Tmem267                                     0   \n",
       "Atg4a                                       0   \n",
       "\n",
       "         ea4352dd-04b6-46ae-aae3-0f1af9ec1b1a  \n",
       "Ulk4                                        0  \n",
       "Smarca2                                     0  \n",
       "Dlgap3                                      0  \n",
       "Tmem267                                     0  \n",
       "Atg4a                                       0  \n",
       "\n",
       "[5 rows x 1089002 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def make_incidence(df):\n",
    "    df['value'] = 1\n",
    "    df['read_index'] = df['read_name'].astype('category').cat.codes\n",
    "    df['gene_index'] = df['gene_name'].astype('category').cat.codes\n",
    "    \n",
    "    data = df['value'].tolist()\n",
    "    row = df['gene_index'].values\n",
    "    col = df['read_index'].values\n",
    "    \n",
    "    n = df['gene_index'].nunique()\n",
    "    m = df['read_index'].nunique()\n",
    "    \n",
    "    index = df['gene_name'].unique()\n",
    "    columns = df['read_name'].unique()\n",
    "    \n",
    "    X = csr_matrix((data, (row, col)), shape=(n, m))\n",
    "    X = csr_matrix((X > 0).astype(int))\n",
    "    H = pd.DataFrame.sparse.from_spmatrix(\n",
    "        X, index=index, columns=columns,\n",
    "    )\n",
    "    del X\n",
    "    del data\n",
    "    del row\n",
    "    del col\n",
    "    del index\n",
    "    del columns\n",
    "    return H\n",
    "\n",
    "\"\"\" FILTER the hyperededges \"\"\"\n",
    "order_threshold = 2.0\n",
    "H = df.copy()\n",
    "H = H[H['is_pt']].reset_index(drop=True)\n",
    "H['order'] = H.groupby('read_name')['gene_name'].transform('nunique')\n",
    "H = H[H['order'] >= order_threshold].reset_index(drop=True)\n",
    "\n",
    "\"\"\"  STRUCTURE the incidence matrix \"\"\"\n",
    "H = make_incidence(H)\n",
    "print(f\"{H.shape=}\")\n",
    "print(f\"{type(H)=}\")\n",
    "print(f\"DataFrame size: {H.memory_usage().sum() / 1024**2:.2f} MB\")\n",
    "\n",
    "outpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/transcription_clusters/protien_coding_incidence_{resolution}.pkl\"\n",
    "H.to_pickle(outpath)\n",
    "H.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13eb6034-8c28-46bb-929b-f5e628800056",
   "metadata": {},
   "source": [
    "# Define and save the core"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ce0be5ac-f25d-4e0b-9d82-1ec565f79750",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(raw) core.shape=(4939267, 19)\n",
      "(grouped) core.shape=(2021658, 6)\n",
      "(filtered) core.shape=(38582, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_name</th>\n",
       "      <th>chroms</th>\n",
       "      <th>tf</th>\n",
       "      <th>core</th>\n",
       "      <th>expressed</th>\n",
       "      <th>mean_expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>0000c863-d9dd-4589-8307-614a9f5db5b1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>7.8320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>0003060b-d552-48f5-a116-6219ff00b425</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>3.2060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>00055dd1-0900-4c4c-986f-2d31b8093830</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6.9375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>00059de3-2ef6-47be-b133-27b1dbb5d212</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>87.3140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>0006c00f-8c61-4d94-946b-8b056e4a292c</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>16.5660</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                read_name  chroms  tf  core  expressed  \\\n",
       "28   0000c863-d9dd-4589-8307-614a9f5db5b1       4   1     2          2   \n",
       "90   0003060b-d552-48f5-a116-6219ff00b425       4   1     2          4   \n",
       "166  00055dd1-0900-4c4c-986f-2d31b8093830       1   1     2          2   \n",
       "180  00059de3-2ef6-47be-b133-27b1dbb5d212       4   3     2          4   \n",
       "217  0006c00f-8c61-4d94-946b-8b056e4a292c       3   1     3          3   \n",
       "\n",
       "     mean_expression  \n",
       "28            7.8320  \n",
       "90            3.2060  \n",
       "166           6.9375  \n",
       "180          87.3140  \n",
       "217          16.5660  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf = 1\n",
    "d = 2\n",
    "t = 2.0\n",
    "\n",
    "\"\"\"criteria:\n",
    "    1. one TF\n",
    "    2. d protien-coding genes in the core with expression over t\n",
    "    3. multiple-chromosomes\n",
    "\"\"\"\n",
    "\n",
    "core = df.copy()\n",
    "print(f\"(raw) {core.shape=}\")\n",
    "core = core[core['expression'].notna()]\n",
    "core = core[core['is_pt']]\n",
    "core['is_expressed'] = core['expression'] > t\n",
    "\n",
    "core = core.groupby('read_name').agg(\n",
    "    chroms = ('chrom', 'nunique'),\n",
    "    tf = ('is_tf', 'sum'),\n",
    "    core = ('is_core', 'sum'),\n",
    "    expressed = ('is_expressed', 'sum'),\n",
    "    mean_expression = ('expression', 'mean'),\n",
    ").reset_index()\n",
    "\n",
    "print(f\"(grouped) {core.shape=}\")\n",
    "\n",
    "core = core[core['tf'] >= tf]\n",
    "core = core[core['core'] >= d]\n",
    "core = core[core['expressed'] >= t]\n",
    "\n",
    "print(f\"(filtered) {core.shape=}\")\n",
    "\n",
    "\n",
    "outpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/transcription_clusters/core_edges_{resolution}.parquet\"\n",
    "core.to_parquet(outpath, index=False)\n",
    "core.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "010ea573-31f9-46ca-8016-b168bc37ddd0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H.shape=(24415, 38582)\n",
      "type(H)=<class 'pandas.core.frame.DataFrame'>\n",
      "DataFrame size: 2.56 MB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>66953ddf-e76d-4cdf-aaf8-be028a2d7b04</th>\n",
       "      <th>c160a170-5af7-412b-9c03-36dfef017384</th>\n",
       "      <th>a8241f5d-1189-49c9-a948-c6a1c91cdf17</th>\n",
       "      <th>3b0686b6-f18f-495d-89f5-8c8b286c2bb1</th>\n",
       "      <th>77bc1796-a0a3-4140-a97b-d4a786d17cb2</th>\n",
       "      <th>4f913e8a-799a-488e-a7bd-6ae9566e5c37</th>\n",
       "      <th>da8a0dfa-5deb-48c3-bf6e-bde5534e0578</th>\n",
       "      <th>a425bdc9-37ea-4020-bc7c-5085fb99a3c7</th>\n",
       "      <th>8ad994fd-c214-46f9-99b0-37c2b3f2946e</th>\n",
       "      <th>b11c9d23-365a-44f5-ace6-217680ce31e8</th>\n",
       "      <th>...</th>\n",
       "      <th>6e74d129-3498-42b0-ade9-0d35f60543af</th>\n",
       "      <th>eae8359f-2057-4492-93d1-10437e892f0b</th>\n",
       "      <th>01f1812f-b821-493c-bc17-3e5e6d453f2c</th>\n",
       "      <th>fcf0a060-2833-4ff0-a352-d5e702f27f46</th>\n",
       "      <th>6199d009-7ef7-44f6-b10c-c0ac846f362c</th>\n",
       "      <th>06857eb1-74b5-484c-9eb3-b0ced3498887</th>\n",
       "      <th>689b7cde-4f15-4869-bda9-bb49cf4e5676</th>\n",
       "      <th>9f9a0a4e-630b-406a-bb24-7026948c9787</th>\n",
       "      <th>acfb51fd-375f-42ac-839c-326616ea1953</th>\n",
       "      <th>de0a50ba-dade-46ba-9ed2-583b92b669b2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Smarca2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sh2d5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ppp1r15b</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dyrk1a</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Igf2bp3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 38582 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          66953ddf-e76d-4cdf-aaf8-be028a2d7b04  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          c160a170-5af7-412b-9c03-36dfef017384  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          a8241f5d-1189-49c9-a948-c6a1c91cdf17  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          3b0686b6-f18f-495d-89f5-8c8b286c2bb1  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          77bc1796-a0a3-4140-a97b-d4a786d17cb2  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          4f913e8a-799a-488e-a7bd-6ae9566e5c37  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          da8a0dfa-5deb-48c3-bf6e-bde5534e0578  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          a425bdc9-37ea-4020-bc7c-5085fb99a3c7  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          8ad994fd-c214-46f9-99b0-37c2b3f2946e  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          b11c9d23-365a-44f5-ace6-217680ce31e8  ...  \\\n",
       "Smarca2                                      0  ...   \n",
       "Sh2d5                                        0  ...   \n",
       "Ppp1r15b                                     0  ...   \n",
       "Dyrk1a                                       0  ...   \n",
       "Igf2bp3                                      0  ...   \n",
       "\n",
       "          6e74d129-3498-42b0-ade9-0d35f60543af  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          eae8359f-2057-4492-93d1-10437e892f0b  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          01f1812f-b821-493c-bc17-3e5e6d453f2c  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          fcf0a060-2833-4ff0-a352-d5e702f27f46  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          6199d009-7ef7-44f6-b10c-c0ac846f362c  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          06857eb1-74b5-484c-9eb3-b0ced3498887  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          689b7cde-4f15-4869-bda9-bb49cf4e5676  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          9f9a0a4e-630b-406a-bb24-7026948c9787  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          acfb51fd-375f-42ac-839c-326616ea1953  \\\n",
       "Smarca2                                      0   \n",
       "Sh2d5                                        0   \n",
       "Ppp1r15b                                     0   \n",
       "Dyrk1a                                       0   \n",
       "Igf2bp3                                      0   \n",
       "\n",
       "          de0a50ba-dade-46ba-9ed2-583b92b669b2  \n",
       "Smarca2                                      0  \n",
       "Sh2d5                                        0  \n",
       "Ppp1r15b                                     0  \n",
       "Dyrk1a                                       0  \n",
       "Igf2bp3                                      0  \n",
       "\n",
       "[5 rows x 38582 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "H = df[df['read_name'].isin(core['read_name'].values)].copy()\n",
    "\n",
    "\"\"\"  STRUCTURE the incidence matrix \"\"\"\n",
    "H = make_incidence(H)\n",
    "print(f\"{H.shape=}\")\n",
    "print(f\"{type(H)=}\")\n",
    "print(f\"DataFrame size: {H.memory_usage().sum() / 1024**2:.2f} MB\")\n",
    "\n",
    "outpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/transcription_clusters/core_incidence_{resolution}.pkl\"\n",
    "H.to_pickle(outpath)\n",
    "H.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5ac854ed-5109-4880-8bfe-f7ddc0cd251b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H.shape=(17186, 34592)\n",
      "type(H)=<class 'pandas.core.frame.DataFrame'>\n",
      "DataFrame size: 2.17 MB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>66953ddf-e76d-4cdf-aaf8-be028a2d7b04</th>\n",
       "      <th>c160a170-5af7-412b-9c03-36dfef017384</th>\n",
       "      <th>3b0686b6-f18f-495d-89f5-8c8b286c2bb1</th>\n",
       "      <th>77bc1796-a0a3-4140-a97b-d4a786d17cb2</th>\n",
       "      <th>4f913e8a-799a-488e-a7bd-6ae9566e5c37</th>\n",
       "      <th>da8a0dfa-5deb-48c3-bf6e-bde5534e0578</th>\n",
       "      <th>a425bdc9-37ea-4020-bc7c-5085fb99a3c7</th>\n",
       "      <th>8ad994fd-c214-46f9-99b0-37c2b3f2946e</th>\n",
       "      <th>2ea6e55f-cc78-418a-b241-f134009153a0</th>\n",
       "      <th>05790af8-be74-4b99-8d1b-49074fa8f81d</th>\n",
       "      <th>...</th>\n",
       "      <th>05fc8d13-3610-4bb0-b173-a908dd526cdd</th>\n",
       "      <th>925b2134-befc-44e1-a9c5-97ca1295c96c</th>\n",
       "      <th>2f5a483f-7f31-4028-b91e-8f8c83a5b922</th>\n",
       "      <th>43cef5b9-05dc-4d4a-b20b-349a91ae2224</th>\n",
       "      <th>49c8ea45-81cb-4f76-a7ad-bd753e8f8c7c</th>\n",
       "      <th>eae8359f-2057-4492-93d1-10437e892f0b</th>\n",
       "      <th>fcf0a060-2833-4ff0-a352-d5e702f27f46</th>\n",
       "      <th>6199d009-7ef7-44f6-b10c-c0ac846f362c</th>\n",
       "      <th>0e705646-2f32-40f9-bad9-fbf5f7ef9d79</th>\n",
       "      <th>9f9a0a4e-630b-406a-bb24-7026948c9787</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Smarca2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sh2d5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dyrk1a</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Igf2bp3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tmem267</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 34592 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         66953ddf-e76d-4cdf-aaf8-be028a2d7b04  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         c160a170-5af7-412b-9c03-36dfef017384  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         3b0686b6-f18f-495d-89f5-8c8b286c2bb1  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         77bc1796-a0a3-4140-a97b-d4a786d17cb2  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         4f913e8a-799a-488e-a7bd-6ae9566e5c37  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         da8a0dfa-5deb-48c3-bf6e-bde5534e0578  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         a425bdc9-37ea-4020-bc7c-5085fb99a3c7  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         8ad994fd-c214-46f9-99b0-37c2b3f2946e  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         2ea6e55f-cc78-418a-b241-f134009153a0  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         05790af8-be74-4b99-8d1b-49074fa8f81d  ...  \\\n",
       "Smarca2                                     0  ...   \n",
       "Sh2d5                                       0  ...   \n",
       "Dyrk1a                                      0  ...   \n",
       "Igf2bp3                                     0  ...   \n",
       "Tmem267                                     0  ...   \n",
       "\n",
       "         05fc8d13-3610-4bb0-b173-a908dd526cdd  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         925b2134-befc-44e1-a9c5-97ca1295c96c  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         2f5a483f-7f31-4028-b91e-8f8c83a5b922  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         43cef5b9-05dc-4d4a-b20b-349a91ae2224  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         49c8ea45-81cb-4f76-a7ad-bd753e8f8c7c  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         eae8359f-2057-4492-93d1-10437e892f0b  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         fcf0a060-2833-4ff0-a352-d5e702f27f46  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         6199d009-7ef7-44f6-b10c-c0ac846f362c  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         0e705646-2f32-40f9-bad9-fbf5f7ef9d79  \\\n",
       "Smarca2                                     0   \n",
       "Sh2d5                                       0   \n",
       "Dyrk1a                                      0   \n",
       "Igf2bp3                                     0   \n",
       "Tmem267                                     0   \n",
       "\n",
       "         9f9a0a4e-630b-406a-bb24-7026948c9787  \n",
       "Smarca2                                     0  \n",
       "Sh2d5                                       0  \n",
       "Dyrk1a                                      0  \n",
       "Igf2bp3                                     0  \n",
       "Tmem267                                     0  \n",
       "\n",
       "[5 rows x 34592 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "order_threshold = 3.0\n",
    "\n",
    "H = df[df['read_name'].isin(core['read_name'].values)].copy()\n",
    "H = H[H['is_pt']].reset_index(drop=True)\n",
    "H['order'] = H.groupby('read_name')['gene_name'].transform('nunique')\n",
    "H = H[H['order'] >= order_threshold].reset_index(drop=True)\n",
    "\n",
    "\n",
    "\"\"\"  STRUCTURE the incidence matrix \"\"\"\n",
    "H = make_incidence(H)\n",
    "print(f\"{H.shape=}\")\n",
    "print(f\"{type(H)=}\")\n",
    "print(f\"DataFrame size: {H.memory_usage().sum() / 1024**2:.2f} MB\")\n",
    "\n",
    "outpath = f\"/scratch/indikar_root/indikar1/shared_data/higher_order/transcription_clusters/core_incidence_{resolution}_protien_coding_only.pkl\"\n",
    "H.to_pickle(outpath)\n",
    "H.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
